diff --git a/src/runtime/runtime-wasm.cc b/src/runtime/runtime-wasm.cc
index 14fa13f667b..07c80038c86 100644
--- a/src/runtime/runtime-wasm.cc
+++ b/src/runtime/runtime-wasm.cc
@@ -612,6 +612,7 @@ RUNTIME_FUNCTION(Runtime_TierUpWasmToJSWrapper) {
     int table_count = trusted_data->dispatch_tables()->length();
     // We have to find the table which contains the correct entry.
     for (int table_index = 0; table_index < table_count; ++table_index) {
+      if (!trusted_data->has_dispatch_table(table_index)) continue;
       Tagged<WasmDispatchTable> table =
           trusted_data->dispatch_table(table_index);
       if (entry_index < table->length() && table->ref(entry_index) == *ref) {
@@ -684,6 +685,7 @@ RUNTIME_FUNCTION(Runtime_TierUpWasmToJSWrapper) {
     int table_count = trusted_data->dispatch_tables()->length();
     // We have to find the table which contains the correct entry.
     for (int table_index = 0; table_index < table_count; ++table_index) {
+      if (!trusted_data->has_dispatch_table(table_index)) continue;
       Tagged<WasmDispatchTable> table =
           trusted_data->dispatch_table(table_index);
       if (entry_index < table->length() && table->ref(entry_index) == *ref) {
diff --git a/src/wasm/wasm-objects-inl.h b/src/wasm/wasm-objects-inl.h
index 37fc7c7efc4..b11a7fe0403 100644
--- a/src/wasm/wasm-objects-inl.h
+++ b/src/wasm/wasm-objects-inl.h
@@ -281,6 +281,12 @@ Tagged<WasmDispatchTable> WasmTrustedInstanceData::dispatch_table(
   return WasmDispatchTable::cast(table);
 }
 
+bool WasmTrustedInstanceData::has_dispatch_table(uint32_t table_index) {
+  Tagged<Object> maybe_table = dispatch_tables()->get(table_index);
+  DCHECK(maybe_table == Smi::zero() || IsWasmDispatchTable(maybe_table));
+  return maybe_table != Smi::zero();
+}
+
 Tagged<WasmModuleObject> WasmTrustedInstanceData::module_object() const {
   return instance_object()->module_object();
 }
diff --git a/src/wasm/wasm-objects.h b/src/wasm/wasm-objects.h
index 0294c51e85d..94961ad5569 100644
--- a/src/wasm/wasm-objects.h
+++ b/src/wasm/wasm-objects.h
@@ -522,6 +522,7 @@ class V8_EXPORT_PRIVATE WasmTrustedInstanceData : public ExposedTrustedObject {
   Address GetCallTarget(uint32_t func_index);
 
   inline Tagged<WasmDispatchTable> dispatch_table(uint32_t table_index);
+  inline bool has_dispatch_table(uint32_t table_index);
 
   // Copies table entries. Returns {false} if the ranges are out-of-bounds.
   static bool CopyTableEntries(
diff --git a/test/mjsunit/regress/wasm/regress-325878101.js b/test/mjsunit/regress/wasm/regress-325878101.js
new file mode 100644
index 00000000000..4e72822a402
--- /dev/null
+++ b/test/mjsunit/regress/wasm/regress-325878101.js
@@ -0,0 +1,26 @@
+// Copyright 2024 the V8 project authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+// Flags: --wasm-wrapper-tiering-budget=1
+
+d8.file.execute('test/mjsunit/wasm/wasm-module-builder.js');
+
+let builder = new WasmModuleBuilder();
+let sig = builder.addType(kSig_v_v);
+let imp = builder.addImport('m', 'f', sig);
+let table0 = builder.addTable(kWasmArrayRef)
+let table1 = builder.addTable(kWasmFuncRef, 1, 1, [kExprRefFunc, 0,])
+builder.addFunction('main', kSig_v_i).exportFunc().addBody([
+  kExprLoop, kWasmVoid,
+  kExprI32Const, 0,
+  kExprCallIndirect, sig, table1.index,
+  kExprLocalGet, 0,
+  kExprI32Const, 1,
+  kExprI32Sub,
+  kExprLocalTee, 0,
+  kExprBrIf, 0,
+  kExprEnd,
+]);
+let instance = builder.instantiate({m: {f: x => x}});
+instance.exports.main(1000);
