diff --git a/src/ast/scopes.cc b/src/ast/scopes.cc
index 62f624025fc..530a249adde 100644
--- a/src/ast/scopes.cc
+++ b/src/ast/scopes.cc
@@ -2693,7 +2693,7 @@ void Scope::AllocateScopeInfosRecursively(
     // next outer scope that needs a context.
     next_outer_scope = scope_info_;
     DCHECK(!scope_info_.is_null());
-    DCHECK(!is_hidden());
+    DCHECK(!is_hidden_catch_scope());
     CHECK_EQ(scope_info_->scope_type(), scope_type_);
     CHECK_EQ(scope_info_->ContextLength(), num_heap_slots_);
 #ifdef DEBUG
@@ -2704,7 +2704,7 @@ void Scope::AllocateScopeInfosRecursively(
 #ifdef DEBUG
     // Mark this ID as being used. Skip hidden scopes because they are
     // synthetic, unreusable, but hard to make unique.
-    if (v8_flags.reuse_scope_infos && !is_hidden()) {
+    if (v8_flags.reuse_scope_infos && !is_hidden_catch_scope()) {
       scope_infos_to_reuse[UniqueIdInScript()] = {};
     }
 #endif
diff --git a/src/ast/scopes.h b/src/ast/scopes.h
index f38b5e3655d..a59c13d0393 100644
--- a/src/ast/scopes.h
+++ b/src/ast/scopes.h
@@ -339,6 +339,10 @@ class V8_EXPORT_PRIVATE Scope : public NON_EXPORTED_BASE(ZoneObject) {
   bool is_hidden() const { return is_hidden_; }
   void set_is_hidden() { is_hidden_ = true; }
 
+  bool is_hidden_catch_scope() const {
+    return is_hidden() && scope_type() == CATCH_SCOPE;
+  }
+
   void ForceContextAllocationForParameters() {
     DCHECK(!already_resolved_);
     force_context_allocation_for_parameters_ = true;
diff --git a/test/mjsunit/regress-352414652.js b/test/mjsunit/regress-352414652.js
new file mode 100644
index 00000000000..d6fb72da915
--- /dev/null
+++ b/test/mjsunit/regress-352414652.js
@@ -0,0 +1,13 @@
+// Copyright 2024 the V8 project authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+// Flags: --allow-natives-syntax --reuse-scope-infos
+
+function __f_11() {
+  for (let i =  __f_26 = function() { return i }; i < 1; ++i) {
+  }
+}
+__f_11();
+  %ForceFlush(__f_11);
+      __f_11();
