diff --git a/src/wasm/wasm-js.cc b/src/wasm/wasm-js.cc
index 67be86a021f..c7ac6d951aa 100644
--- a/src/wasm/wasm-js.cc
+++ b/src/wasm/wasm-js.cc
@@ -1261,8 +1261,19 @@ void WebAssemblyInstantiateImpl(
             data->Get(1).As<Value>().As<Promise::Resolver>();
         Local<Value> imports = data->Get(2).As<Value>();
 
-        i::DirectHandle<i::WasmModuleObject> module_obj =
-            i::Cast<i::WasmModuleObject>(Utils::OpenDirectHandle(*info[0]));
+        i::DirectHandle<i::WasmModuleObject> module_obj;
+        if (!TryCast(Utils::OpenDirectHandle(*info[0]), &module_obj)) {
+          auto callback = reinterpret_cast<i::Isolate*>(isolate2)
+                              ->wasm_async_resolve_promise_callback();
+          CHECK(callback);
+          Local<Value> error = Exception::TypeError(String::NewFromUtf8Literal(
+              isolate2,
+              "Illegal call to WebAssembly.instantiate callback: Argument is "
+              "not a WasmModuleObject"));
+          callback(isolate2, context, promise_resolver, error,
+                   WasmAsyncSuccess::kFail);
+          return;
+        }
 
         i::wasm::GetWasmEngine()->AsyncInstantiate(
             reinterpret_cast<i::Isolate*>(isolate2),
diff --git a/test/mjsunit/regress/wasm/regress-463089250.js b/test/mjsunit/regress/wasm/regress-463089250.js
new file mode 100644
index 00000000000..ab7d3613c0f
--- /dev/null
+++ b/test/mjsunit/regress/wasm/regress-463089250.js
@@ -0,0 +1,15 @@
+// Copyright 2025 the V8 project authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+d8.file.execute('test/mjsunit/wasm/wasm-module-builder.js');
+
+const builder = new WasmModuleBuilder();
+WebAssembly.Module.prototype.then = resolve => resolve(17);
+
+assertPromiseResult(
+    WebAssembly.instantiate(builder.toBuffer()), assertUnreachable,
+    e => assertException(
+        e, TypeError,
+        'Illegal call to WebAssembly.instantiate callback: ' +
+            'Argument is not a WasmModuleObject'));
