diff --git a/src/objects/js-atomics-synchronization.cc b/src/objects/js-atomics-synchronization.cc
index 9e03d07812b..78c5de5326b 100644
--- a/src/objects/js-atomics-synchronization.cc
+++ b/src/objects/js-atomics-synchronization.cc
@@ -955,7 +955,24 @@ void JSAtomicsMutex::UnlockAsyncLockedMutex(
           async_locked_waiter_wrapper->foreign_address<kWaiterQueueForeignTag>(
               IsolateForSandbox(requester)));
   LockAsyncWaiterQueueNode::RemoveFromAsyncWaiterQueueList(waiter_node);
-  Unlock(requester);
+  if (IsCurrentThreadOwner()) {
+    Unlock(requester);
+    return;
+  }
+  // If this is reached, the lock was already released by this thread.
+  // This can happen if waitAsync is called without awaiting or due to
+  // promise prototype tampering. Setting Promise.prototype.then to a
+  // non callable will cause the `waiting_for_callback_promise` (defined in
+  // LockOrEnqueuePromise) reactions to be called even if the async callback
+  // is not resolved; as a consequence, the following code will try to unlock
+  // the mutex twice:
+  //
+  // let mutex = new Atomics.Mutex();
+  // let cv = new Atomics.Condition();
+  // Promise.prototype.then = undefined;
+  // Atomics.Mutex.lockAsync(mutex, async function() {
+  //   await Atomics.Condition.waitAsync(cv, mutex);
+  // }
 }
 
 bool JSAtomicsMutex::DequeueTimedOutAsyncWaiter(
diff --git a/test/mjsunit/shared-memory/condition-async-wait-no-await.js b/test/mjsunit/shared-memory/condition-async-wait-no-await.js
new file mode 100644
index 00000000000..66095656fa7
--- /dev/null
+++ b/test/mjsunit/shared-memory/condition-async-wait-no-await.js
@@ -0,0 +1,27 @@
+// Copyright 2024 the V8 project authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+//
+// Flags: --harmony-struct --allow-natives-syntax
+
+(function TestAsyncWaitPromisePrototypeTampering() {
+  const mutex = new Atomics.Mutex;
+  const cv = new Atomics.Condition;
+  let finishedExecution = false;
+  let lockAsyncPromise = Atomics.Mutex.lockAsync(mutex, async () => {
+    Atomics.Condition.waitAsync(cv, mutex);
+    finishedExecution = true;
+  });
+
+  let thenExecuted = false;
+  lockAsyncPromise.then(() => {
+    thenExecuted = true;
+  });
+
+  // Run Microtasks
+  setTimeout(() => {
+    assertTrue(finishedExecution);
+    assertTrue(thenExecuted);
+    assertEquals(1, %AtomicsSynchronizationPrimitiveNumWaitersForTesting(cv));
+  }, 0);
+})();
diff --git a/test/mjsunit/shared-memory/regress-crbug-371659887.js b/test/mjsunit/shared-memory/regress-crbug-371659887.js
new file mode 100644
index 00000000000..4f7146090f5
--- /dev/null
+++ b/test/mjsunit/shared-memory/regress-crbug-371659887.js
@@ -0,0 +1,40 @@
+// Copyright 2024 the V8 project authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+//
+// Flags: --harmony-struct
+
+// If Promise.prototype.then is set to a non-callable, builtin
+// promises become non thenables, hence any promise resolve to
+// them calls its reactions instead of queing a then task.
+// In the current implementation an internal promise with
+// a reaction to release the lock is resolved to the result of
+// the asyncLock callback, causing the unlock microtask to
+// be executed before the asyncLock callback finishes. This test
+// should be updated if the behavior intentionally changes.
+(function TestAsyncWaitPromisePrototypeTampering() {
+  const mutex = new Atomics.Mutex;
+  const cv = new Atomics.Condition;
+  let finishedExecution = false;
+  let startedExecution = false;
+  let lockAsyncPromise = Atomics.Mutex.lockAsync(mutex, async () => {
+    startedExecution = true;
+    await Atomics.Condition.waitAsync(cv, mutex);
+    finishedExecution = true;
+  });
+
+  let thenExecuted = false;
+  lockAsyncPromise.then(() => {
+    thenExecuted = true;
+  });
+
+  Promise.prototype.then = undefined;
+  // Run Microtasks
+  setTimeout(() => {
+    assertTrue(startedExecution);
+    assertFalse(finishedExecution);
+    // The lockAsyncPromise is resolved even though the async callback
+    // didn't finish executing.
+    assertTrue(thenExecuted);
+  }, 0);
+})();
