diff --git a/src/flags/flag-definitions.h b/src/flags/flag-definitions.h
index cc8dde9386a..08f6e612a10 100644
--- a/src/flags/flag-definitions.h
+++ b/src/flags/flag-definitions.h
@@ -2393,8 +2393,9 @@ DEFINE_BOOL_READONLY(fast_map_update, false,
 DEFINE_INT(max_valid_polymorphic_map_count, 4,
            "maximum number of valid maps to track in POLYMORPHIC state")
 DEFINE_BOOL(
-    clone_object_sidestep_transitions, true,
+    clone_object_sidestep_transitions, false,
     "support sidestep transitions for dependency tracking object clone maps")
+DEFINE_WEAK_IMPLICATION(future, move_prototype_transitions_first)
 
 // map-inl.h
 DEFINE_INT(fast_properties_soft_limit, 12,
