diff --git a/src/base/logging.h b/src/base/logging.h
index ea7b08ae07c..94fac83d21e 100644
--- a/src/base/logging.h
+++ b/src/base/logging.h
@@ -423,6 +423,12 @@ DEFINE_SIGNED_MISMATCH_COMP(is_unsigned_vs_signed, GE, CmpLEImpl(rhs, lhs))
 #undef MAKE_UNSIGNED
 #undef DEFINE_SIGNED_MISMATCH_COMP
 
+// For CHECK_BOUNDS, define to-unsigned conversion helpers.
+template <typename T>
+constexpr std::make_unsigned_t<T> ToUnsigned(T val) {
+  return static_cast<std::make_unsigned_t<T>>(val);
+}
+
 // Helper functions for CHECK_OP macro.
 // The (float, float) and (double, double) instantiations are explicitly
 // externalized to ensure proper 32/64-bit comparisons on x86.
@@ -466,6 +472,10 @@ DEFINE_CHECK_OP_IMPL(GT, > )
 #define CHECK_NOT_NULL(val) CHECK((val) != nullptr)
 #define CHECK_IMPLIES(lhs, rhs) \
   CHECK_WITH_MSG(!(lhs) || (rhs), #lhs " implies " #rhs)
+// Performs a single (unsigned) comparison to check that {index} is
+// in range [0, limit).
+#define CHECK_BOUNDS(index, limit) \
+  CHECK_LT(v8::base::ToUnsigned(index), v8::base::ToUnsigned(limit))
 
 }  // namespace base
 }  // namespace v8
@@ -484,6 +494,8 @@ DEFINE_CHECK_OP_IMPL(GT, > )
 #define DCHECK_NOT_NULL(val) DCHECK((val) != nullptr)
 #define DCHECK_IMPLIES(lhs, rhs) \
   DCHECK_WITH_MSG(!(lhs) || (rhs), #lhs " implies " #rhs)
+#define DCHECK_BOUNDS(index, limit) \
+  DCHECK_LT(v8::base::ToUnsigned(index), v8::base::ToUnsigned(limit))
 #else
 #define DCHECK(condition)      ((void) 0)
 #define DCHECK_WITH_LOC(condition, location) ((void)0)
@@ -497,6 +509,7 @@ DEFINE_CHECK_OP_IMPL(GT, > )
 #define DCHECK_NULL(val)       ((void) 0)
 #define DCHECK_NOT_NULL(val)   ((void) 0)
 #define DCHECK_IMPLIES(v1, v2) ((void) 0)
+#define DCHECK_BOUNDS(index, limit) ((void)0)
 #endif
 
 #endif  // V8_BASE_LOGGING_H_
diff --git a/src/execution/arguments.h b/src/execution/arguments.h
index 22ce32bbb0f..d3a156229cc 100644
--- a/src/execution/arguments.h
+++ b/src/execution/arguments.h
@@ -74,6 +74,8 @@ class Arguments {
     // Corruption of certain heap objects (see e.g. crbug.com/1507223) can lead
     // to OOB arguments access, and therefore OOB stack access. This SBXCHECK
     // defends against that.
+    // Note: "LE" is intentional: it's okay to compute the address of the
+    // first nonexistent entry.
     SBXCHECK_LE(static_cast<uint32_t>(index), static_cast<uint32_t>(length_));
     uintptr_t offset = index * kSystemPointerSize;
     if (arguments_type == ArgumentsType::kJS) {
diff --git a/src/regexp/experimental/experimental-interpreter.cc b/src/regexp/experimental/experimental-interpreter.cc
index 9a46267f7e0..30bd24325c3 100644
--- a/src/regexp/experimental/experimental-interpreter.cc
+++ b/src/regexp/experimental/experimental-interpreter.cc
@@ -615,8 +615,7 @@ class NfaInterpreter {
     DCHECK_GT(clock, 0);
 
     while (true) {
-      SBXCHECK_GE(t.pc, 0);
-      SBXCHECK_LT(t.pc, bytecode_.length());
+      SBXCHECK_BOUNDS(t.pc, bytecode_.size());
       if (IsPcProcessed(t.pc, t.consumed_since_last_quantifier)) {
         DestroyThread(t);
         return RegExp::kInternalRegExpSuccess;
@@ -694,8 +693,8 @@ class NfaInterpreter {
           active_threads_.Rewind(0);
           return RegExp::kInternalRegExpSuccess;
         case RegExpInstruction::SET_REGISTER_TO_CP:
-          SBXCHECK_GE(inst.payload.register_index, 0);
-          SBXCHECK_LT(inst.payload.register_index, register_count_per_match_);
+          SBXCHECK_BOUNDS(inst.payload.register_index,
+                          register_count_per_match_);
           GetRegisterArray(t)[inst.payload.register_index] = input_index_;
           if (v8_flags.experimental_regexp_engine_capture_group_opt) {
             GetCaptureClockArray(t)[inst.payload.register_index] = clock;
@@ -703,8 +702,8 @@ class NfaInterpreter {
           ++t.pc;
           break;
         case RegExpInstruction::CLEAR_REGISTER:
-          SBXCHECK_GE(inst.payload.register_index, 0);
-          SBXCHECK_LT(inst.payload.register_index, register_count_per_match_);
+          SBXCHECK_BOUNDS(inst.payload.register_index,
+                          register_count_per_match_);
           GetRegisterArray(t)[inst.payload.register_index] =
               kUndefinedRegisterValue;
           ++t.pc;
@@ -738,8 +737,8 @@ class NfaInterpreter {
           // Reaching this instruction means that the current lookbehind thread
           // has found a match and needs to be destroyed. Since the lookbehind
           // is verified at this position, we update the `lookbehind_table_`.
-          SBXCHECK_GE(inst.payload.looktable_index, 0);
-          SBXCHECK_LT(inst.payload.looktable_index, lookbehind_table_.length());
+          SBXCHECK_BOUNDS(inst.payload.looktable_index,
+                          lookbehind_table_.length());
           lookbehind_table_[inst.payload.looktable_index] = true;
           DestroyThread(t);
           return RegExp::kInternalRegExpSuccess;
@@ -751,8 +750,7 @@ class NfaInterpreter {
           // position.
           const int32_t lookbehind_index =
               inst.payload.read_lookbehind.lookbehind_index();
-          SBXCHECK_GE(lookbehind_index, 0);
-          SBXCHECK_LT(lookbehind_index, lookbehind_table_.length());
+          SBXCHECK_BOUNDS(lookbehind_index, lookbehind_table_.length());
           if (lookbehind_table_[lookbehind_index] !=
               inst.payload.read_lookbehind.is_positive()) {
             DestroyThread(t);
diff --git a/src/sandbox/check.h b/src/sandbox/check.h
index 05a306d1434..c2d7b82ffd9 100644
--- a/src/sandbox/check.h
+++ b/src/sandbox/check.h
@@ -50,6 +50,7 @@
 #define SBXCHECK_GE(lhs, rhs) SBXCHECK_WRAPPED(GE, lhs, rhs)
 #define SBXCHECK_LT(lhs, rhs) SBXCHECK_WRAPPED(LT, lhs, rhs)
 #define SBXCHECK_LE(lhs, rhs) SBXCHECK_WRAPPED(LE, lhs, rhs)
+#define SBXCHECK_BOUNDS(index, limit) SBXCHECK_WRAPPED(BOUNDS, index, limit)
 #else
 #define SBXCHECK(condition) DCHECK(condition)
 #define SBXCHECK_EQ(lhs, rhs) DCHECK_EQ(lhs, rhs)
@@ -58,6 +59,7 @@
 #define SBXCHECK_GE(lhs, rhs) DCHECK_GE(lhs, rhs)
 #define SBXCHECK_LT(lhs, rhs) DCHECK_LT(lhs, rhs)
 #define SBXCHECK_LE(lhs, rhs) DCHECK_LE(lhs, rhs)
+#define SBXCHECK_BOUNDS(index, limit) DCHECK_BOUNDS(index, limit)
 #endif
 
 #endif  // V8_SANDBOX_CHECK_H_
diff --git a/src/wasm/wasm-objects.cc b/src/wasm/wasm-objects.cc
index a615b847b83..382ce1b4a7b 100644
--- a/src/wasm/wasm-objects.cc
+++ b/src/wasm/wasm-objects.cc
@@ -306,7 +306,7 @@ void WasmTableObject::SetFunctionTableEntry(Isolate* isolate,
         exported_function->instance_data(), isolate);
     int func_index = exported_function->function_index();
     const WasmModule* module = target_instance_data->module();
-    SBXCHECK_LT(func_index, module->functions.size());
+    SBXCHECK_BOUNDS(func_index, module->functions.size());
     auto* wasm_function = module->functions.data() + func_index;
     UpdateDispatchTables(isolate, table, entry_index, wasm_function,
                          target_instance_data);
@@ -1364,7 +1364,7 @@ void WasmTrustedInstanceData::InitDataSegmentArrays(
 
 Address WasmTrustedInstanceData::GetCallTarget(uint32_t func_index) {
   wasm::NativeModule* native_module = this->native_module();
-  SBXCHECK_LT(func_index, native_module->num_functions());
+  SBXCHECK_BOUNDS(func_index, native_module->num_functions());
   if (func_index < native_module->num_imported_functions()) {
     return dispatch_table_for_imports()->target(func_index);
   }
@@ -1940,7 +1940,7 @@ void WasmDispatchTable::Set(int index, Tagged<Object> ref, Address call_target,
     return;
   }
 
-  SBXCHECK_LT(index, length());
+  SBXCHECK_BOUNDS(index, length());
   DCHECK(IsWasmApiFunctionRef(ref) || IsWasmTrustedInstanceData(ref));
   DCHECK_EQ(ref == Smi::zero(), call_target == kNullAddress);
   const int offset = OffsetOf(index);
@@ -1953,7 +1953,7 @@ void WasmDispatchTable::Set(int index, Tagged<Object> ref, Address call_target,
 
 void WasmDispatchTable::SetForImport(int index, Tagged<TrustedObject> ref,
                                      Address call_target) {
-  SBXCHECK_LT(index, length());
+  SBXCHECK_BOUNDS(index, length());
   DCHECK(IsWasmApiFunctionRef(ref) || IsWasmTrustedInstanceData(ref));
   DCHECK_NE(kNullAddress, call_target);
   const int offset = OffsetOf(index);
@@ -1966,7 +1966,7 @@ void WasmDispatchTable::SetForImport(int index, Tagged<TrustedObject> ref,
 }
 
 void WasmDispatchTable::Clear(int index) {
-  SBXCHECK_LT(index, length());
+  SBXCHECK_BOUNDS(index, length());
   const int offset = OffsetOf(index);
   ClearProtectedPointerField(offset + kRefBias);
   WriteField<Address>(offset + kTargetBias, kNullAddress);
@@ -1974,7 +1974,7 @@ void WasmDispatchTable::Clear(int index) {
 }
 
 void WasmDispatchTable::SetTarget(int index, Address call_target) {
-  SBXCHECK_LT(index, length());
+  SBXCHECK_BOUNDS(index, length());
   const int offset = OffsetOf(index) + kTargetBias;
   WriteField<Address>(offset, call_target);
 }
diff --git a/test/mjsunit/mjsunit.status b/test/mjsunit/mjsunit.status
index 5eb53e2f203..6828b5f5bb0 100644
--- a/test/mjsunit/mjsunit.status
+++ b/test/mjsunit/mjsunit.status
@@ -463,6 +463,7 @@
   'turboshaft/maglev-frontend/has-in-prototype-chain-wasm': [SKIP],
 
   'sandbox/regress/regress-334120897': [SKIP],
+  'sandbox/regress/regress-349502157': [SKIP],
   'sandbox/wasm-inlining-sigcheck': [SKIP],
   'sandbox/wasm-signature-verification': [SKIP],
   'sandbox/wasm-table-import': [SKIP],
diff --git a/test/mjsunit/sandbox/regress/regress-349502157.js b/test/mjsunit/sandbox/regress/regress-349502157.js
new file mode 100644
index 00000000000..85dcd9a0ae8
--- /dev/null
+++ b/test/mjsunit/sandbox/regress/regress-349502157.js
@@ -0,0 +1,70 @@
+// Copyright 2024 the V8 project authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+// Flags: --sandbox-testing
+
+d8.file.execute('test/mjsunit/wasm/wasm-module-builder.js');
+
+let builder = new WasmModuleBuilder();
+let $struct = builder.addStruct([makeField(kWasmI64, true)]);
+let $sig_v_ls = builder.addType(makeSig([kWasmI64, wasmRefType($struct)], []));
+let $sig_v_ll = builder.addType(makeSig([kWasmI64, kWasmI64], []));
+let $writer = builder.addFunction("writer", $sig_v_ls)
+  .exportFunc()
+  .addBody([
+    kExprLocalGet, 1,
+    kExprLocalGet, 0,
+    kGCPrefix, kExprStructSet, $struct, 0,
+  ]);
+let $boom = builder.addFunction("boom", $sig_v_ll)
+  .exportFunc()
+  .addBody([
+    kExprLocalGet, 1,
+    kExprLocalGet, 0,
+    kExprI32Const, 0,
+    kExprCallIndirect, $sig_v_ll, 0,
+  ])
+let $dummy = builder.addFunction("dummy", $sig_v_ll).exportFunc().addBody([]);
+// Target table.
+let $t0 = builder
+    .addTable(wasmRefType($sig_v_ll), 1, 1, [kExprRefFunc, $dummy.index])
+    .exportAs('table_v_ll');
+// Padding tables for alignment.
+let $td0 = builder.addTable(
+    wasmRefType($sig_v_ll), 1, 1, [kExprRefFunc, $dummy.index]);
+let $td1 = builder.addTable(
+    wasmRefType($sig_v_ll), 1, 1, [kExprRefFunc, $dummy.index]);
+let $td2 = builder.addTable(
+    wasmRefType($sig_v_ll), 1, 1, [kExprRefFunc, $dummy.index]);
+// OOB table.
+let $t1 = builder
+    .addTable(wasmRefType($sig_v_ls), 1, 1, [kExprRefFunc, $writer.index])
+    .exportAs("table_v_ls");
+
+let instance = builder.instantiate();
+let { writer, dummy, boom, table_v_ls, table_v_ll } = instance.exports;
+
+// Prepare corruption utilities.
+const kHeapObjectTag = 1;
+const kWasmTableObjectCurrentLengthOffset = 16;
+const kWasmTableObjectMaximumLengthOffset = 20;
+let memory = new DataView(new Sandbox.MemoryView(0, 0x100000000));
+function getPtr(obj) {
+  return Sandbox.getAddressOf(obj) + kHeapObjectTag;
+}
+function getField(obj, offset) {
+  return memory.getUint32(obj + offset - kHeapObjectTag, true);
+}
+function setField(obj, offset, value) {
+  memory.setUint32(obj + offset - kHeapObjectTag, value, true);
+}
+
+const kSmiMinusOne = 0xfffffffe;
+setField(getPtr(table_v_ls), kWasmTableObjectCurrentLengthOffset, kSmiMinusOne);
+setField(getPtr(table_v_ls), kWasmTableObjectMaximumLengthOffset, kSmiMinusOne);
+
+// Check bypassed, write @ index -7 -> writes into table_v_ll dispatch table!
+table_v_ls.set(0xfffffff9, writer);
+
+boom(BigInt(Sandbox.targetPage) - 0x7n, 42n);
