diff --git a/src/compiler/js-call-reducer.cc b/src/compiler/js-call-reducer.cc
index d17942a75cc..8bd5a058da2 100644
--- a/src/compiler/js-call-reducer.cc
+++ b/src/compiler/js-call-reducer.cc
@@ -5172,6 +5172,18 @@ TNode<Object> JSCallReducerAssembler::ReduceJSCallWithArrayLikeOrSpreadOfEmpty(
   DCHECK_EQ(static_cast<Node*>(arguments_list)->opcode(),
             IrOpcode::kJSCreateEmptyLiteralArray);
 
+  // Check that arguments_list's prototype is still an array prototype.
+  TNode<Map> map = LoadMap(TNode<HeapObject>::UncheckedCast(arguments_list));
+  TNode<HeapObject> proto = TNode<HeapObject>::UncheckedCast(
+      LoadField(AccessBuilder::ForMapPrototype(), map));
+  TNode<HeapObject> initial_array_prototype =
+      HeapConstant(broker()
+                       ->target_native_context()
+                       .initial_array_prototype(broker())
+                       .object());
+  TNode<Boolean> check = ReferenceEqual(proto, initial_array_prototype);
+  CheckIf(check, DeoptimizeReason::kWrongMap, p.feedback());
+
   // Turn the JSCallWithArrayLike or JSCallWithSpread roughly into:
   //
   //      "arguments_list array is still empty?"
diff --git a/test/mjsunit/compiler/regress-385386138.js b/test/mjsunit/compiler/regress-385386138.js
new file mode 100644
index 00000000000..72dadea5ff5
--- /dev/null
+++ b/test/mjsunit/compiler/regress-385386138.js
@@ -0,0 +1,27 @@
+// Copyright 2024 the V8 project authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+// Flags: --allow-natives-syntax --turbofan --no-always-turbofan
+
+function foo() {
+  let val = [];
+  val.__proto__ = RegExp();
+  return Math.max(...val);
+}
+
+%PrepareFunctionForOptimization(foo);
+assertThrows(() => foo(), TypeError,
+             "Spread syntax requires ...iterable[Symbol.iterator] to be a function");
+
+%OptimizeFunctionOnNextCall(foo);
+assertThrows(() => foo(), TypeError,
+             "Spread syntax requires ...iterable[Symbol.iterator] to be a function");
+assertUnoptimized(foo);
+
+
+%OptimizeFunctionOnNextCall(foo);
+assertThrows(() => foo(), TypeError,
+             "Spread syntax requires ...iterable[Symbol.iterator] to be a function");
+// TF should not have speculatively optimized CallWithSpread.
+assertOptimized(foo);
