diff --git a/src/objects/js-number-format.cc b/src/objects/js-number-format.cc
index d47185bed04..742416ceee8 100644
--- a/src/objects/js-number-format.cc
+++ b/src/objects/js-number-format.cc
@@ -1482,6 +1482,14 @@ MaybeDirectHandle<JSNumberFormat> JSNumberFormat::New(
 
 namespace {
 
+template <typename StringHandle>
+int32_t SignedStringLength(StringHandle string) {
+  uint32_t unsigned_length = string->length();
+  static_assert(String::kMaxLength < std::numeric_limits<int32_t>::max());
+  SBXCHECK_LE(unsigned_length, String::kMaxLength);
+  return static_cast<int32_t>(unsigned_length);
+}
+
 icu::number::FormattedNumber FormatDecimalString(
     Isolate* isolate,
     const icu::number::LocalizedNumberFormatter& number_format,
@@ -1489,7 +1497,7 @@ icu::number::FormattedNumber FormatDecimalString(
   string = String::Flatten(isolate, string);
   DisallowGarbageCollection no_gc;
   const String::FlatContent& flat = string->GetFlatContent(no_gc);
-  int32_t length = static_cast<int32_t>(string->length());
+  int32_t length = SignedStringLength(string);
   if (flat.IsOneByte()) {
     const char* char_buffer =
         reinterpret_cast<const char*>(flat.ToOneByteVector().begin());
@@ -1532,7 +1540,7 @@ Maybe<icu::number::FormattedNumber> IcuFormatNumber(
     big_int_string = String::Flatten(isolate, big_int_string);
     DisallowGarbageCollection no_gc;
     const String::FlatContent& flat = big_int_string->GetFlatContent(no_gc);
-    int32_t length = static_cast<int32_t>(big_int_string->length());
+    int32_t length = SignedStringLength(big_int_string);
     DCHECK(flat.IsOneByte());
     const char* char_buffer =
         reinterpret_cast<const char*>(flat.ToOneByteVector().begin());
@@ -1545,7 +1553,7 @@ Maybe<icu::number::FormattedNumber> IcuFormatNumber(
           String::Flatten(isolate, Cast<String>(numeric_obj));
       DisallowGarbageCollection no_gc;
       const String::FlatContent& flat = string->GetFlatContent(no_gc);
-      int32_t length = static_cast<int32_t>(string->length());
+      int32_t length = SignedStringLength(string);
       if (flat.IsOneByte()) {
         const char* char_buffer =
             reinterpret_cast<const char*>(flat.ToOneByteVector().begin());
@@ -1782,7 +1790,7 @@ Maybe<icu::Formattable> IntlMathematicalValue::ToFormattable(
   {
     DisallowGarbageCollection no_gc;
     const String::FlatContent& flat = string->GetFlatContent(no_gc);
-    int32_t length = static_cast<int32_t>(string->length());
+    int32_t length = SignedStringLength(string);
     if (flat.IsOneByte()) {
       icu::Formattable result(
           {reinterpret_cast<const char*>(flat.ToOneByteVector().begin()),
