diff --git a/src/objects/map.cc b/src/objects/map.cc
index 3856abba0e0..db276027352 100644
--- a/src/objects/map.cc
+++ b/src/objects/map.cc
@@ -1279,7 +1279,7 @@ Handle<Map> Map::Normalize(Isolate* isolate, Handle<Map> fast_map,
       Handle<Map> fresh = Map::CopyNormalized(isolate, fast_map, mode);
       fresh->set_elements_kind(new_elements_kind);
       if (!new_prototype.is_null()) {
-        fresh->set_prototype(*new_prototype);
+        Map::SetPrototype(isolate, fresh, new_prototype);
       }
 
       static_assert(Map::kPrototypeValidityCellOffset ==
@@ -1321,7 +1321,8 @@ Handle<Map> Map::Normalize(Isolate* isolate, Handle<Map> fast_map,
     new_map = Map::CopyNormalized(isolate, fast_map, mode);
     new_map->set_elements_kind(new_elements_kind);
     if (!new_prototype.is_null()) {
-      new_map->set_prototype(*new_prototype);
+      Map::SetPrototype(isolate, new_map, new_prototype);
+      DCHECK(new_map->is_dictionary_map() && !new_map->is_deprecated());
     }
     if (use_cache) {
       cache->Set(fast_map, new_map);
diff --git a/test/mjsunit/regress/regress-341972220.js b/test/mjsunit/regress/regress-341972220.js
new file mode 100644
index 00000000000..a99a65a66cb
--- /dev/null
+++ b/test/mjsunit/regress/regress-341972220.js
@@ -0,0 +1,13 @@
+// Copyright 2024 the V8 project authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+class __c_1 extends Function {
+  constructor(...__v_23) {
+    super(... __v_23);
+  }
+}
+// Create obj which triggers normalization
+var __v_19 = new __c_1("'use strict';");
+// Set prototype to a r/o object
+__v_19.__proto__ = [];
