diff --git a/src/compiler/js-context-specialization.cc b/src/compiler/js-context-specialization.cc
index 8d7fc651207..cabaac0322f 100644
--- a/src/compiler/js-context-specialization.cc
+++ b/src/compiler/js-context-specialization.cc
@@ -258,7 +258,7 @@ Reduction JSContextSpecialization::ReduceJSStoreScriptContext(Node* node) {
     // The value is not a constant any more, so we don't need to generate
     // code for invalidating the side data.
     const Operator* op =
-        jsgraph_->javascript()->StoreContext(access.depth(), access.index());
+        jsgraph_->javascript()->StoreContext(depth, access.index());
     Node* new_store = effect = jsgraph_->graph()->NewNode(
         op, NodeProperties::GetValueInput(node, 0), context, effect, control);
     ReplaceWithValue(node, new_store, effect, control);
diff --git a/test/mjsunit/regress-377352368.js b/test/mjsunit/regress-377352368.js
new file mode 100644
index 00000000000..30baf0fb6b5
--- /dev/null
+++ b/test/mjsunit/regress-377352368.js
@@ -0,0 +1,25 @@
+// Copyright 2024 the V8 project authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+//
+// Flags: --allow-natives-syntax
+
+let o0 = {};
+function f1() {
+    function f2() {
+    }
+    function F4( a8) {
+        function f10() {
+            a8 = f2;
+        }
+        o0 = a8;
+    }
+    %PrepareFunctionForOptimization(f2);
+    %PrepareFunctionForOptimization(F4);
+    new F4();
+}
+%PrepareFunctionForOptimization(f1);
+f1();
+f1();
+%OptimizeFunctionOnNextCall(f1);
+f1();
