diff --git a/src/maglev/maglev-graph-builder.cc b/src/maglev/maglev-graph-builder.cc
index 3abbaa5c5af..28be7271264 100644
--- a/src/maglev/maglev-graph-builder.cc
+++ b/src/maglev/maglev-graph-builder.cc
@@ -4853,8 +4853,10 @@ ValueNode* MaglevGraphBuilder::BuildNumberOrOddballToFloat64(
 ReduceResult MaglevGraphBuilder::BuildCheckSmi(ValueNode* object,
                                                bool elidable) {
   if (CheckStaticType(object, NodeType::kSmi)) return object;
-  if (CheckType(object, NodeType::kAnyHeapObject)) {
-    return EmitUnconditionalDeopt(DeoptimizeReason::kNotASmi);
+  // Check for the empty type first so that we catch the case where
+  // GetType(object) is already empty.
+  if (IsEmptyNodeType(CombineType(GetType(object), NodeType::kSmi))) {
+    return EmitUnconditionalDeopt(DeoptimizeReason::kSmi);
   }
   if (EnsureType(object, NodeType::kSmi) && elidable) return object;
   switch (object->value_representation()) {
@@ -4881,21 +4883,26 @@ ReduceResult MaglevGraphBuilder::BuildCheckSmi(ValueNode* object,
 }
 
 ReduceResult MaglevGraphBuilder::BuildCheckHeapObject(ValueNode* object) {
-  if (EnsureType(object, NodeType::kAnyHeapObject)) return ReduceResult::Done();
+  // Check for the empty type first so that we catch the case where
+  // GetType(object) is already empty.
   if (IsEmptyNodeType(CombineType(GetType(object), NodeType::kAnyHeapObject))) {
     return EmitUnconditionalDeopt(DeoptimizeReason::kSmi);
   }
+  if (EnsureType(object, NodeType::kAnyHeapObject)) return ReduceResult::Done();
   AddNewNode<CheckHeapObject>({object});
   return ReduceResult::Done();
 }
 
 ReduceResult MaglevGraphBuilder::BuildCheckString(ValueNode* object) {
   NodeType known_type;
-  if (EnsureType(object, NodeType::kString, &known_type))
-    return ReduceResult::Done();
+  // Check for the empty type first so that we catch the case where
+  // GetType(object) is already empty.
   if (IsEmptyNodeType(CombineType(GetType(object), NodeType::kString))) {
     return EmitUnconditionalDeopt(DeoptimizeReason::kNotAString);
   }
+  if (EnsureType(object, NodeType::kString, &known_type)) {
+    return ReduceResult::Done();
+  }
   AddNewNode<CheckString>({object}, GetCheckType(known_type));
   return ReduceResult::Done();
 }
@@ -4903,43 +4910,51 @@ ReduceResult MaglevGraphBuilder::BuildCheckString(ValueNode* object) {
 ReduceResult MaglevGraphBuilder::BuildCheckStringOrStringWrapper(
     ValueNode* object) {
   NodeType known_type;
-  if (EnsureType(object, NodeType::kStringOrStringWrapper, &known_type))
-    return ReduceResult::Done();
+  // Check for the empty type first so that we catch the case where
+  // GetType(object) is already empty.
   if (IsEmptyNodeType(
           CombineType(GetType(object), NodeType::kStringOrStringWrapper))) {
     return EmitUnconditionalDeopt(DeoptimizeReason::kNotAStringOrStringWrapper);
   }
+  if (EnsureType(object, NodeType::kStringOrStringWrapper, &known_type))
+    return ReduceResult::Done();
   AddNewNode<CheckStringOrStringWrapper>({object}, GetCheckType(known_type));
   return ReduceResult::Done();
 }
 
 ReduceResult MaglevGraphBuilder::BuildCheckNumber(ValueNode* object) {
-  if (EnsureType(object, NodeType::kNumber)) return ReduceResult::Done();
+  // Check for the empty type first so that we catch the case where
+  // GetType(object) is already empty.
   if (IsEmptyNodeType(CombineType(GetType(object), NodeType::kNumber))) {
     return EmitUnconditionalDeopt(DeoptimizeReason::kNotANumber);
   }
+  if (EnsureType(object, NodeType::kNumber)) return ReduceResult::Done();
   AddNewNode<CheckNumber>({object}, Object::Conversion::kToNumber);
   return ReduceResult::Done();
 }
 
 ReduceResult MaglevGraphBuilder::BuildCheckSymbol(ValueNode* object) {
   NodeType known_type;
-  if (EnsureType(object, NodeType::kSymbol, &known_type))
-    return ReduceResult::Done();
+  // Check for the empty type first so that we catch the case where
+  // GetType(object) is already empty.
   if (IsEmptyNodeType(CombineType(GetType(object), NodeType::kSymbol))) {
     return EmitUnconditionalDeopt(DeoptimizeReason::kNotASymbol);
   }
+  if (EnsureType(object, NodeType::kSymbol, &known_type))
+    return ReduceResult::Done();
   AddNewNode<CheckSymbol>({object}, GetCheckType(known_type));
   return ReduceResult::Done();
 }
 
 ReduceResult MaglevGraphBuilder::BuildCheckJSReceiver(ValueNode* object) {
   NodeType known_type;
-  if (EnsureType(object, NodeType::kJSReceiver, &known_type))
-    return ReduceResult::Done();
+  // Check for the empty type first so that we catch the case where
+  // GetType(object) is already empty.
   if (IsEmptyNodeType(CombineType(GetType(object), NodeType::kJSReceiver))) {
     return EmitUnconditionalDeopt(DeoptimizeReason::kWrongInstanceType);
   }
+  if (EnsureType(object, NodeType::kJSReceiver, &known_type))
+    return ReduceResult::Done();
   AddNewNode<CheckInstanceType>({object}, GetCheckType(known_type),
                                 FIRST_JS_RECEIVER_TYPE, LAST_JS_RECEIVER_TYPE);
   return ReduceResult::Done();
@@ -4948,14 +4963,16 @@ ReduceResult MaglevGraphBuilder::BuildCheckJSReceiver(ValueNode* object) {
 ReduceResult MaglevGraphBuilder::BuildCheckJSReceiverOrNullOrUndefined(
     ValueNode* object) {
   NodeType known_type;
-  if (EnsureType(object, NodeType::kJSReceiverOrNullOrUndefined, &known_type)) {
-    return ReduceResult::Done();
-  }
+  // Check for the empty type first so that we catch the case where
+  // GetType(object) is already empty.
   if (IsEmptyNodeType(CombineType(GetType(object),
                                   NodeType::kJSReceiverOrNullOrUndefined))) {
     return EmitUnconditionalDeopt(
         DeoptimizeReason::kNotAJavaScriptObjectOrNullOrUndefined);
   }
+  if (EnsureType(object, NodeType::kJSReceiverOrNullOrUndefined, &known_type)) {
+    return ReduceResult::Done();
+  }
   AddNewNode<CheckJSReceiverOrNullOrUndefined>({object},
                                                GetCheckType(known_type));
   return ReduceResult::Done();
diff --git a/src/maglev/maglev-ir.h b/src/maglev/maglev-ir.h
index da41426b145..b966cf90e37 100644
--- a/src/maglev/maglev-ir.h
+++ b/src/maglev/maglev-ir.h
@@ -776,6 +776,10 @@ inline bool IsInstanceOfNodeType(compiler::MapRef map, NodeType type,
 }
 
 inline std::ostream& operator<<(std::ostream& out, const NodeType& type) {
+  if (IsEmptyNodeType(type)) {
+    out << "Empty";
+    return out;
+  }
   switch (type) {
 #define CASE(Name, _)     \
   case NodeType::k##Name: \
diff --git a/test/mjsunit/maglev/regress-408571498.js b/test/mjsunit/maglev/regress-408571498.js
new file mode 100644
index 00000000000..dee9e157a3f
--- /dev/null
+++ b/test/mjsunit/maglev/regress-408571498.js
@@ -0,0 +1,16 @@
+// Copyright 2025 the V8 project authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+// Flags: --allow-natives-syntax --maglev --no-always-turbofan
+// Flags: --maglev-cons-string-elision
+
+const v0 = [4.5];
+function f7() {
+  v0[268435] = undefined;
+  `foooo000000000000000000000000000 ${undefined}`;
+}
+%PrepareFunctionForOptimization(f7);
+f7();
+%OptimizeMaglevOnNextCall(f7);
+f7();
