diff --git a/src/compiler/turboshaft/late-load-elimination-reducer.cc b/src/compiler/turboshaft/late-load-elimination-reducer.cc
index 5f40949b965..a0f2940d4b3 100644
--- a/src/compiler/turboshaft/late-load-elimination-reducer.cc
+++ b/src/compiler/turboshaft/late-load-elimination-reducer.cc
@@ -356,16 +356,22 @@ void LateLoadEliminationAnalyzer::ProcessStore(OpIndex op_idx,
     return;
   }
 
-  OpIndex value = store.value();
-
   // Updating the known stored values.
   if (!invalidate_maybe_aliasing) memory_.Invalidate(store);
   memory_.Insert(store);
 
   // Updating aliases if the value stored was known as non-aliasing.
+  OpIndex value = store.value();
   if (non_aliasing_objects_.HasKeyFor(value)) {
     non_aliasing_objects_.Set(value, false);
   }
+
+  // If we just stored a map, invalidate the maps for this base.
+  if (store.offset == HeapObject::kMapOffset && !store.index().valid()) {
+    if (object_maps_.HasKeyFor(store.base())) {
+      object_maps_.Set(store.base(), MapMaskAndOr{});
+    }
+  }
 }
 
 // Since we only loosely keep track of what can or can't alias, we assume that
diff --git a/test/mjsunit/compiler/regress-340663085.js b/test/mjsunit/compiler/regress-340663085.js
new file mode 100644
index 00000000000..ce2c77a0915
--- /dev/null
+++ b/test/mjsunit/compiler/regress-340663085.js
@@ -0,0 +1,29 @@
+// Copyright 2024 the V8 project authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+function f(b) {
+  var v1 = {};
+  v1.prop0 = 1;
+
+  var v2 = {};
+  v2.prop0 = 1;
+
+  if (b) {
+    v2 = v1;
+  }
+
+  (function (v3) {
+    v3.prop0;
+    v2.a = 11;
+    v2.b = 12;
+    v3.x = 21;
+    v2.y = 22;
+  })(v1);
+
+  return v2.x;
+}
+
+assertEquals(undefined, f());
+assertEquals(21, f(true));
+assertEquals(21, f(true));
