diff --git a/src/maglev/maglev-ir.h b/src/maglev/maglev-ir.h
index 7e038ddd68c..85339091d0c 100644
--- a/src/maglev/maglev-ir.h
+++ b/src/maglev/maglev-ir.h
@@ -5005,6 +5005,11 @@ class CheckedNumberToInt32
   void PrintParams(std::ostream&) const {}
 };
 
+// TODO(dmercadier): Number->Float64 conversions are exact and could still be
+// marked with the ConversionNode OpProperties, but Oddball->Float64 are not,
+// which is why this node doesn't have the ConversionNode property. We could
+// consider splitting this node in 2 (or for Numbers, one for NumberOrOddball),
+// as well as adding a "tonumber_float64" alternative.
 class UncheckedNumberOrOddballToFloat64
     : public FixedInputValueNodeT<1, UncheckedNumberOrOddballToFloat64> {
   using Base = FixedInputValueNodeT<1, UncheckedNumberOrOddballToFloat64>;
@@ -5015,8 +5020,7 @@ class UncheckedNumberOrOddballToFloat64
       : Base(TaggedToFloat64ConversionTypeOffset::update(bitfield,
                                                          conversion_type)) {}
 
-  static constexpr OpProperties kProperties =
-      OpProperties::Float64() | OpProperties::ConversionNode();
+  static constexpr OpProperties kProperties = OpProperties::Float64();
   static constexpr
       typename Base::InputTypes kInputTypes{ValueRepresentation::kTagged};
 
diff --git a/test/mjsunit/maglev/regress-449784529.js b/test/mjsunit/maglev/regress-449784529.js
new file mode 100644
index 00000000000..4ae6ec3d120
--- /dev/null
+++ b/test/mjsunit/maglev/regress-449784529.js
@@ -0,0 +1,16 @@
+// Copyright 2025 the V8 project authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+// Flags: --allow-natives-syntax
+
+function foo(x, y) {
+  return (y < x) - (x < x);
+}
+
+%PrepareFunctionForOptimization(foo);
+assertEquals(0, foo(9));
+assertEquals(0, foo(9));
+
+%OptimizeMaglevOnNextCall(foo);
+assertEquals(0, foo(9));
