diff --git a/src/compiler/turboshaft/maglev-graph-building-phase.cc b/src/compiler/turboshaft/maglev-graph-building-phase.cc
index 10b83b54385..bb5ea4fc3f6 100644
--- a/src/compiler/turboshaft/maglev-graph-building-phase.cc
+++ b/src/compiler/turboshaft/maglev-graph-building-phase.cc
@@ -2726,7 +2726,12 @@ class GraphBuilder {
             node->eager_deopt_info()->feedback_to_update()));
     return maglev::ProcessResult::kContinue;
   }
-  maglev::ProcessResult Process(maglev::UnsafeSmiTag* node,
+  maglev::ProcessResult Process(maglev::UnsafeSmiTagInt32* node,
+                                const maglev::ProcessingState& state) {
+    SetMap(node, __ TagSmi(Map(node->input())));
+    return maglev::ProcessResult::kContinue;
+  }
+  maglev::ProcessResult Process(maglev::UnsafeSmiTagUint32* node,
                                 const maglev::ProcessingState& state) {
     SetMap(node, __ TagSmi(Map(node->input())));
     return maglev::ProcessResult::kContinue;
diff --git a/src/maglev/maglev-graph-builder.cc b/src/maglev/maglev-graph-builder.cc
index 1a788d1814b..dc6e7d6204e 100644
--- a/src/maglev/maglev-graph-builder.cc
+++ b/src/maglev/maglev-graph-builder.cc
@@ -1449,13 +1449,13 @@ ValueNode* MaglevGraphBuilder::GetTaggedValue(
   switch (representation) {
     case ValueRepresentation::kInt32: {
       if (NodeTypeIsSmi(node_info->type())) {
-        return alternative.set_tagged(AddNewNode<UnsafeSmiTag>({value}));
+        return alternative.set_tagged(AddNewNode<UnsafeSmiTagInt32>({value}));
       }
       return alternative.set_tagged(AddNewNode<Int32ToNumber>({value}));
     }
     case ValueRepresentation::kUint32: {
       if (NodeTypeIsSmi(node_info->type())) {
-        return alternative.set_tagged(AddNewNode<UnsafeSmiTag>({value}));
+        return alternative.set_tagged(AddNewNode<UnsafeSmiTagUint32>({value}));
       }
       return alternative.set_tagged(AddNewNode<Uint32ToNumber>({value}));
     }
@@ -1506,13 +1506,13 @@ ReduceResult MaglevGraphBuilder::GetSmiValue(
   switch (representation) {
     case ValueRepresentation::kInt32: {
       if (NodeTypeIsSmi(node_info->type())) {
-        return alternative.set_tagged(AddNewNode<UnsafeSmiTag>({value}));
+        return alternative.set_tagged(AddNewNode<UnsafeSmiTagInt32>({value}));
       }
       return alternative.set_tagged(AddNewNode<CheckedSmiTagInt32>({value}));
     }
     case ValueRepresentation::kUint32: {
       if (NodeTypeIsSmi(node_info->type())) {
-        return alternative.set_tagged(AddNewNode<UnsafeSmiTag>({value}));
+        return alternative.set_tagged(AddNewNode<UnsafeSmiTagUint32>({value}));
       }
       return alternative.set_tagged(AddNewNode<CheckedSmiTagUint32>({value}));
     }
@@ -3464,7 +3464,8 @@ NodeType StaticTypeForNode(compiler::JSHeapBroker* broker,
     case Opcode::kCheckedSmiTagInt32:
     case Opcode::kCheckedSmiTagUint32:
     case Opcode::kCheckedSmiTagFloat64:
-    case Opcode::kUnsafeSmiTag:
+    case Opcode::kUnsafeSmiTagInt32:
+    case Opcode::kUnsafeSmiTagUint32:
     case Opcode::kSmiConstant:
       return NodeType::kSmi;
     case Opcode::kInt32ToNumber:
diff --git a/src/maglev/maglev-graph-builder.h b/src/maglev/maglev-graph-builder.h
index 01156bb528c..a404d51203e 100644
--- a/src/maglev/maglev-graph-builder.h
+++ b/src/maglev/maglev-graph-builder.h
@@ -9,6 +9,7 @@
 #include <iomanip>
 #include <map>
 #include <type_traits>
+#include <utility>
 
 #include "src/base/functional.h"
 #include "src/base/logging.h"
@@ -211,7 +212,7 @@ class MaglevGraphBuilder {
     // Don't use the AddNewNode helper for the function entry stack check, so
     // that we can set a custom deopt frame on it.
     FunctionEntryStackCheck* function_entry_stack_check =
-        NodeBase::New<FunctionEntryStackCheck>(zone(), {});
+        NodeBase::New<FunctionEntryStackCheck>(zone(), 0);
     new (function_entry_stack_check->lazy_deopt_info()) LazyDeoptInfo(
         zone(), GetDeoptFrameForEntryStackCheck(),
         interpreter::Register::invalid_value(), 0, compiler::FeedbackSource());
@@ -910,8 +911,9 @@ class MaglevGraphBuilder {
         known_node_aspects().available_expressions.erase(exists);
       }
     }
-    NodeT* node =
-        NodeBase::New<NodeT>(zone(), inputs, std::forward<Args>(args)...);
+    NodeT* node = NodeBase::New<NodeT>(zone(), inputs.size(),
+                                       std::forward<Args>(args)...);
+    SetNodeInputs(node, inputs);
     DCHECK_EQ(node->options(), std::tuple{std::forward<Args>(args)...});
     uint32_t epoch = Node::needs_epoch_check(op)
                          ? known_node_aspects().effect_epoch()
@@ -943,8 +945,10 @@ class MaglevGraphBuilder {
         }
       }
     }
-    NodeT* node =
-        NodeBase::New<NodeT>(zone(), inputs, std::forward<Args>(args)...);
+    static_assert(IsFixedInputNode<NodeT>());
+    NodeT* node = NodeBase::New<NodeT>(zone(), inputs.size(),
+                                       std::forward<Args>(args)...);
+    SetNodeInputs(node, inputs);
     return AttachExtraInfoAndAddToGraph(node);
   }
 
@@ -1106,7 +1110,15 @@ class MaglevGraphBuilder {
           },
           kBuiltin, GetContext());
     } else {
-      return AddNewNode<CallBuiltin>(inputs, kBuiltin);
+      return AddNewNode<CallBuiltin>(
+          inputs.size(),
+          [&](CallBuiltin* call_builtin) {
+            int arg_index = 0;
+            for (auto* input : inputs) {
+              call_builtin->set_arg(arg_index++, input);
+            }
+          },
+          kBuiltin);
     }
   }
 
@@ -1676,11 +1688,43 @@ class MaglevGraphBuilder {
     PrintVirtualObjects();
   }
 
+  ValueNode* ConvertInputTo(ValueNode* input, ValueRepresentation expected) {
+    ValueRepresentation repr = input->properties().value_representation();
+    if (repr == expected) return input;
+    switch (expected) {
+      case ValueRepresentation::kTagged:
+        return GetTaggedValue(input);
+      case ValueRepresentation::kInt32:
+        return GetInt32(input);
+      case ValueRepresentation::kFloat64:
+      case ValueRepresentation::kHoleyFloat64:
+        return GetFloat64(input);
+      case ValueRepresentation::kUint32:
+      case ValueRepresentation::kIntPtr:
+        // These conversion should be explicitly done beforehand.
+        UNREACHABLE();
+    }
+  }
+
+  template <typename NodeT>
+  void SetNodeInputs(NodeT* node, std::initializer_list<ValueNode*> inputs) {
+    // Nodes with zero input count don't have kInputTypes defined.
+    if constexpr (NodeT::kInputCount > 0) {
+      int i = 0;
+      for (ValueNode* input : inputs) {
+        DCHECK_NOT_NULL(input);
+        node->set_input(i, ConvertInputTo(input, NodeT::kInputTypes[i]));
+        i++;
+      }
+    }
+  }
+
   template <typename ControlNodeT, typename... Args>
   BasicBlock* FinishBlock(std::initializer_list<ValueNode*> control_inputs,
                           Args&&... args) {
     ControlNodeT* control_node = NodeBase::New<ControlNodeT>(
-        zone(), control_inputs, std::forward<Args>(args)...);
+        zone(), control_inputs.size(), std::forward<Args>(args)...);
+    SetNodeInputs(control_node, control_inputs);
     AttachEagerDeoptInfo(control_node);
     AttachDeoptCheckpoint(control_node);
     static_assert(!ControlNodeT::kProperties.can_lazy_deopt());
diff --git a/src/maglev/maglev-interpreter-frame-state.cc b/src/maglev/maglev-interpreter-frame-state.cc
index 2771f37fce7..23d1e686940 100644
--- a/src/maglev/maglev-interpreter-frame-state.cc
+++ b/src/maglev/maglev-interpreter-frame-state.cc
@@ -442,10 +442,10 @@ ValueNode* FromInt32ToTagged(const MaglevGraphBuilder* builder,
              value->Is<BuiltinStringPrototypeCharCodeOrCodePointAt>()) {
     static_assert(String::kMaxLength <= kSmiMaxValue,
                   "String length must fit into a Smi");
-    tagged = Node::New<UnsafeSmiTag>(builder->zone(), {value});
+    tagged = Node::New<UnsafeSmiTagInt32>(builder->zone(), {value});
   } else if (NodeTypeIsSmi(node_type)) {
     // For known Smis, we can tag without a check.
-    tagged = Node::New<UnsafeSmiTag>(builder->zone(), {value});
+    tagged = Node::New<UnsafeSmiTagInt32>(builder->zone(), {value});
   } else {
     tagged = Node::New<Int32ToNumber>(builder->zone(), {value});
   }
@@ -464,7 +464,7 @@ ValueNode* FromUint32ToTagged(const MaglevGraphBuilder* builder,
 
   ValueNode* tagged;
   if (NodeTypeIsSmi(node_type)) {
-    tagged = Node::New<UnsafeSmiTag>(builder->zone(), {value});
+    tagged = Node::New<UnsafeSmiTagUint32>(builder->zone(), {value});
   } else {
     tagged = Node::New<Uint32ToNumber>(builder->zone(), {value});
   }
diff --git a/src/maglev/maglev-ir.cc b/src/maglev/maglev-ir.cc
index 1359a44b760..b70ed257e89 100644
--- a/src/maglev/maglev-ir.cc
+++ b/src/maglev/maglev-ir.cc
@@ -618,35 +618,12 @@ void CheckValueInputIs(const NodeBase* node, int i, Opcode expected,
   }
 }
 
-void CheckValueInputIsWord32(const NodeBase* node, int i,
-                             MaglevGraphLabeller* graph_labeller) {
-  ValueNode* input = node->input(i).node();
-  DCHECK(!input->Is<Identity>());
-  ValueRepresentation got = input->properties().value_representation();
-  if (got != ValueRepresentation::kInt32 &&
-      got != ValueRepresentation::kUint32) {
-    std::ostringstream str;
-    str << "Type representation error: node ";
-    if (graph_labeller) {
-      str << "#" << graph_labeller->NodeId(node) << " : ";
-    }
-    str << node->opcode() << " (input @" << i << " = " << input->opcode()
-        << ") type " << got << " is not Word32 (Int32 or Uint32)";
-    FATAL("%s", str.str().c_str());
-  }
-}
-
 void GeneratorStore::VerifyInputs(MaglevGraphLabeller* graph_labeller) const {
   for (int i = 0; i < input_count(); i++) {
     CheckValueInputIs(this, i, ValueRepresentation::kTagged, graph_labeller);
   }
 }
 
-void UnsafeSmiTag::VerifyInputs(MaglevGraphLabeller* graph_labeller) const {
-  DCHECK_EQ(input_count(), 1);
-  CheckValueInputIsWord32(this, 0, graph_labeller);
-}
-
 void Phi::VerifyInputs(MaglevGraphLabeller* graph_labeller) const {
   switch (value_representation()) {
 #define CASE_REPR(repr)                                        \
@@ -1562,23 +1539,22 @@ void CheckedSmiTagUint32::GenerateCode(MaglevAssembler* masm,
   __ SmiTagUint32AndJumpIfFail(reg, fail);
 }
 
-void UnsafeSmiTag::SetValueLocationConstraints() {
+void UnsafeSmiTagInt32::SetValueLocationConstraints() {
   UseRegister(input());
   DefineSameAsFirst(this);
 }
-void UnsafeSmiTag::GenerateCode(MaglevAssembler* masm,
-                                const ProcessingState& state) {
-  Register reg = ToRegister(input());
-  switch (input().node()->properties().value_representation()) {
-    case ValueRepresentation::kInt32:
-      __ UncheckedSmiTagInt32(reg);
-      break;
-    case ValueRepresentation::kUint32:
-      __ UncheckedSmiTagUint32(reg);
-      break;
-    default:
-      UNREACHABLE();
-  }
+void UnsafeSmiTagInt32::GenerateCode(MaglevAssembler* masm,
+                                     const ProcessingState& state) {
+  __ UncheckedSmiTagInt32(ToRegister(input()));
+}
+
+void UnsafeSmiTagUint32::SetValueLocationConstraints() {
+  UseRegister(input());
+  DefineSameAsFirst(this);
+}
+void UnsafeSmiTagUint32::GenerateCode(MaglevAssembler* masm,
+                                      const ProcessingState& state) {
+  __ UncheckedSmiTagUint32(ToRegister(input()));
 }
 
 void CheckedSmiIncrement::SetValueLocationConstraints() {
diff --git a/src/maglev/maglev-ir.h b/src/maglev/maglev-ir.h
index ced824bb795..9025f31be59 100644
--- a/src/maglev/maglev-ir.h
+++ b/src/maglev/maglev-ir.h
@@ -217,7 +217,8 @@ class MergePointInterpreterFrameState;
   V(CheckedSmiSizedInt32)                           \
   V(CheckedSmiTagInt32)                             \
   V(CheckedSmiTagUint32)                            \
-  V(UnsafeSmiTag)                                   \
+  V(UnsafeSmiTagInt32)                              \
+  V(UnsafeSmiTagUint32)                             \
   V(CheckedSmiUntag)                                \
   V(UnsafeSmiUntag)                                 \
   V(CheckedInternalizedString)                      \
@@ -1673,6 +1674,7 @@ class NodeBase : public ZoneObject {
   template <class Derived, typename... Args>
   static Derived* New(Zone* zone, std::initializer_list<ValueNode*> inputs,
                       Args&&... args) {
+    static_assert(Derived::kProperties.is_conversion());
     Derived* node =
         Allocate<Derived>(zone, inputs.size(), std::forward<Args>(args)...);
 
@@ -1705,6 +1707,8 @@ class NodeBase : public ZoneObject {
     bitfield_ = OpPropertiesField::update(bitfield_, properties);
   }
 
+  inline void set_input(int index, ValueNode* node);
+
   template <class T>
   constexpr bool Is() const;
 
@@ -1919,7 +1923,6 @@ class NodeBase : public ZoneObject {
   }
 
   inline void initialize_input_null(int index);
-  inline void set_input(int index, ValueNode* node);
 
   // For nodes that don't have data past the input, allow trimming the input
   // count. This is used by Phis to reduce inputs when merging in dead control
@@ -2506,9 +2509,9 @@ struct YouNeedToDefineAnInputTypesArrayInYourDerivedClass {};
 // properties), and known numbers of inputs.
 template <size_t InputCount, typename Base, typename Derived>
 class FixedInputNodeTMixin : public NodeTMixin<Base, Derived> {
+ public:
   static constexpr size_t kInputCount = InputCount;
 
- public:
   // Shadowing for static knowledge.
   constexpr bool has_inputs() const { return input_count() > 0; }
   constexpr uint16_t input_count() const { return kInputCount; }
@@ -2554,6 +2557,12 @@ class FixedInputNodeTMixin : public NodeTMixin<Base, Derived> {
   }
 };
 
+template <class T, class = void>
+struct IsFixedInputNode : public std::false_type {};
+template <class T>
+struct IsFixedInputNode<T, std::void_t<decltype(T::kInputCount)>>
+    : public std::true_type {};
+
 template <class Derived>
 using NodeT = NodeTMixin<Node, Derived>;
 
@@ -3207,17 +3216,41 @@ class CheckedSmiTagUint32
 };
 
 // Input must guarantee to fit in a Smi.
-class UnsafeSmiTag : public FixedInputValueNodeT<1, UnsafeSmiTag> {
-  using Base = FixedInputValueNodeT<1, UnsafeSmiTag>;
+class UnsafeSmiTagInt32 : public FixedInputValueNodeT<1, UnsafeSmiTagInt32> {
+  using Base = FixedInputValueNodeT<1, UnsafeSmiTagInt32>;
 
  public:
-  explicit UnsafeSmiTag(uint64_t bitfield) : Base(bitfield) {}
+  explicit UnsafeSmiTagInt32(uint64_t bitfield) : Base(bitfield) {}
 
   static constexpr OpProperties kProperties = OpProperties::ConversionNode();
+  static constexpr
+      typename Base::InputTypes kInputTypes{ValueRepresentation::kInt32};
 
   Input& input() { return Node::input(0); }
 
-  void VerifyInputs(MaglevGraphLabeller*) const;
+#ifdef V8_COMPRESS_POINTERS
+  void MarkTaggedInputsAsDecompressing() {
+    // No tagged inputs.
+  }
+#endif
+
+  void SetValueLocationConstraints();
+  void GenerateCode(MaglevAssembler*, const ProcessingState&);
+  void PrintParams(std::ostream&, MaglevGraphLabeller*) const {}
+};
+
+// Input must guarantee to fit in a Smi.
+class UnsafeSmiTagUint32 : public FixedInputValueNodeT<1, UnsafeSmiTagUint32> {
+  using Base = FixedInputValueNodeT<1, UnsafeSmiTagUint32>;
+
+ public:
+  explicit UnsafeSmiTagUint32(uint64_t bitfield) : Base(bitfield) {}
+
+  static constexpr OpProperties kProperties = OpProperties::ConversionNode();
+  static constexpr
+      typename Base::InputTypes kInputTypes{ValueRepresentation::kUint32};
+
+  Input& input() { return Node::input(0); }
 
 #ifdef V8_COMPRESS_POINTERS
   void MarkTaggedInputsAsDecompressing() {
@@ -8460,6 +8493,9 @@ class CallWithArrayLike : public FixedInputValueNodeT<4, CallWithArrayLike> {
   explicit CallWithArrayLike(uint64_t bitfield) : Base(bitfield) {}
 
   static constexpr OpProperties kProperties = OpProperties::JSCall();
+  static constexpr typename Base::InputTypes kInputTypes{
+      ValueRepresentation::kTagged, ValueRepresentation::kTagged,
+      ValueRepresentation::kTagged, ValueRepresentation::kTagged};
 
   Input& function() { return input(kFunctionIndex); }
   Input& receiver() { return input(kReceiverIndex); }
diff --git a/src/maglev/maglev-regalloc.cc b/src/maglev/maglev-regalloc.cc
index 6fbb94fa542..37954cbd845 100644
--- a/src/maglev/maglev-regalloc.cc
+++ b/src/maglev/maglev-regalloc.cc
@@ -1163,7 +1163,7 @@ void StraightForwardRegisterAllocator::AddMoveBeforeCurrentNode(
           << PrintNodeLabel(graph_labeller(), node) << std::endl;
     }
     gap_move =
-        Node::New<ConstantGapMove>(compilation_info_->zone(), {}, node, target);
+        Node::New<ConstantGapMove>(compilation_info_->zone(), 0, node, target);
   } else {
     if (v8_flags.trace_maglev_regalloc) {
       printing_visitor_->os() << "  gap move: " << target << " ← "
@@ -1171,7 +1171,7 @@ void StraightForwardRegisterAllocator::AddMoveBeforeCurrentNode(
                               << source << std::endl;
     }
     gap_move =
-        Node::New<GapMove>(compilation_info_->zone(), {},
+        Node::New<GapMove>(compilation_info_->zone(), 0,
                            compiler::AllocatedOperand::cast(source), target);
   }
   if (compilation_info_->has_graph_labeller()) {
diff --git a/test/mjsunit/maglev/regress-349864879.js b/test/mjsunit/maglev/regress-349864879.js
new file mode 100644
index 00000000000..bf7d198347e
--- /dev/null
+++ b/test/mjsunit/maglev/regress-349864879.js
@@ -0,0 +1,113 @@
+// Copyright 2024 the V8 project authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+//
+// Flags: --allow-natives-syntax
+
+(function() {
+  function f0() {
+    const v3 = ([-639548.3633151249,-1e-15]).with();
+    const v4 = ("NffG").charCodeAt(f0);
+    const v5 = v4 - v4;
+    function f6(a7, a8, a9) {
+        try { a7.reduce(); } catch (e) {}
+        return a7;
+    }
+    %PrepareFunctionForOptimization(f6);
+    f6(v5);
+    f6(v3);
+    return f0;
+  }
+  %PrepareFunctionForOptimization(f0);
+  const v13 = f0();
+  v13.call(v13, v13);
+  %OptimizeMaglevOnNextCall(f0);
+  f0();
+})();
+
+(function() {
+  function foo(v1) {
+    const v2 = v1 % v1;
+    function f3(a4) {
+        a4.x;
+    }
+    %PrepareFunctionForOptimization(f3);
+    f3(v2);
+    f3(-9007199254740990n);
+  }
+  %PrepareFunctionForOptimization(foo);
+  foo(0);
+  foo(1);
+  %OptimizeMaglevOnNextCall(foo);
+  foo(2);
+})();
+
+(function() {
+  const v1 = new Uint16Array(Uint16Array, Uint16Array, Uint16Array);
+  function f2() {
+      const o4 = {
+          "d": 1563771623,
+      };
+      return o4;
+  }
+  const v5 = f2();
+  function f6() {
+      function f7(a8, a9, a10, a11) {
+          a8.d.toString();
+          return a8;
+      }
+      %PrepareFunctionForOptimization(f7);
+      f7(v5);
+      f7(f6);
+      return f2;
+  }
+  function f16() {
+      return v1;
+  }
+  Object.defineProperty(f6, "d", { get: f16 });
+  %PrepareFunctionForOptimization(f6);
+  f6();
+  f6();
+  %OptimizeMaglevOnNextCall(f6);
+  f6();
+})();
+
+(function() {
+  %PrepareFunctionForOptimization(f3);
+  %PrepareFunctionForOptimization(foo);
+  function F0() {}
+  const v2 = new F0();
+  function f3(a4) {
+    return a4.b;
+  }
+  f3(v2);
+  const v10 = new Float64Array(256);
+  function foo(v11) {
+    const v12 = v11.substring();
+    const v13 = Math.min();
+    const v14 = v13 - v13;
+    v12 > v10;
+    f3(v14);
+  }
+  foo("42");
+  foo("42");
+  %OptimizeMaglevOnNextCall(foo);
+  foo("42");
+})();
+
+(function() {
+  const v2 = [1.1,,];
+  function f3(a4) {
+      const v5 = v2[a4];
+      if (v2[a4] === undefined) {
+          return 1;
+      }
+      v5.toString();
+      return undefined;
+  }
+  %PrepareFunctionForOptimization(f3);
+  f3(0);
+  f3(1);
+  %OptimizeMaglevOnNextCall(f3);
+  f3(v2);
+})();
