diff --git a/src/compiler/turboshaft/maglev-graph-building-phase.cc b/src/compiler/turboshaft/maglev-graph-building-phase.cc
index 069171c5b78..faf2d2aebb4 100644
--- a/src/compiler/turboshaft/maglev-graph-building-phase.cc
+++ b/src/compiler/turboshaft/maglev-graph-building-phase.cc
@@ -4750,6 +4750,12 @@ base::Optional<BailoutReason> MaglevGraphBuildingPhase::Run(PipelineData* data,
       compilation_info->toplevel_compilation_unit(), &bailout);
   builder.ProcessGraph(maglev_graph);
 
+  // Copying {inlined_functions} from Maglev to Turboshaft.
+  for (OptimizedCompilationInfo::InlinedFunctionHolder holder :
+       maglev_graph->inlined_functions()) {
+    data->info()->inlined_functions().push_back(holder);
+  }
+
   return bailout;
 }
 
diff --git a/test/mjsunit/turboshaft/maglev-frontend/regress-351418008.js b/test/mjsunit/turboshaft/maglev-frontend/regress-351418008.js
new file mode 100644
index 00000000000..800a425d69b
--- /dev/null
+++ b/test/mjsunit/turboshaft/maglev-frontend/regress-351418008.js
@@ -0,0 +1,22 @@
+// Copyright 2024 the V8 project authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+// Flags: --expose-gc --allow-natives-syntax --no-lazy-feedback-allocation
+// Flags: --turboshaft-from-maglev --stress-flush-code
+
+let callGC = function () {
+  gc();
+};
+
+function f() {
+  const o = {};
+  for (let i = 0; i <= 1; ++i) {
+    callGC();
+    o.my_property = 2;  // Polymorphic store.
+  }
+}
+%PrepareFunctionForOptimization(f);
+f();
+%OptimizeFunctionOnNextCall(f);
+f();
