diff --git a/src/maglev/maglev-phi-representation-selector.cc b/src/maglev/maglev-phi-representation-selector.cc
index 312ba7632dc..938e44f27d8 100644
--- a/src/maglev/maglev-phi-representation-selector.cc
+++ b/src/maglev/maglev-phi-representation-selector.cc
@@ -1164,21 +1164,19 @@ void MaglevPhiRepresentationSelector::PreparePhiTaggings(
 
     // We create a Phi to merge all of the existing taggings.
     int predecessor_count = new_block->predecessor_count();
-    Phi* phi = reducer_.AddNewNode<Phi>(
-        predecessor_count,
-        [&](Phi* phi) {
-          for (int i = 0; static_cast<size_t>(i) < predecessors.size(); i++) {
-            phi->set_input(i, predecessors[i]);
-          }
-          if (predecessors.size() != static_cast<size_t>(predecessor_count)) {
-            // The backedge is omitted from {predecessors}. With set the Phi as
-            // its own backedge.
-            DCHECK(new_block->is_loop());
-            phi->set_input(predecessor_count - 1, phi);
-          }
-          new_block->AddPhi(phi);
-        },
-        new_block->state(), interpreter::Register());
+    Phi* phi = Node::New<Phi>(zone(), predecessor_count, new_block->state(),
+                              interpreter::Register());
+    for (int i = 0; static_cast<size_t>(i) < predecessors.size(); i++) {
+      phi->set_input(i, predecessors[i]);
+    }
+    if (predecessors.size() != static_cast<size_t>(predecessor_count)) {
+      // The backedge is omitted from {predecessors}. With set the Phi as its
+      // own backedge.
+      DCHECK(new_block->is_loop());
+      phi->set_input(predecessor_count - 1, phi);
+    }
+    reducer_.RegisterNode(phi);
+    new_block->AddPhi(phi);
 
     return phi;
   };
diff --git a/test/mjsunit/regress/regress-428226995.js b/test/mjsunit/regress/regress-428226995.js
new file mode 100644
index 00000000000..d9f0c9c8392
--- /dev/null
+++ b/test/mjsunit/regress/regress-428226995.js
@@ -0,0 +1,27 @@
+// Copyright 2025 the V8 project authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+// Flags: --allow-natives-syntax
+
+function foo(opt) {
+  class C {}
+  function bar() {
+    let x = 0;
+    for (let i = 0; i < 5; i++) {
+      C[2] = x;
+      C[2] = x;
+      x++;
+    }
+  }
+  %PrepareFunctionForOptimization(bar);
+  bar();
+  bar();
+  if (opt) {
+    %OptimizeMaglevOnNextCall(bar);
+  }
+  bar();
+}
+
+foo(false);
+foo(true)
