diff --git a/src/ic/ic.cc b/src/ic/ic.cc
index 1c44bcfc049..dfa63ffc399 100644
--- a/src/ic/ic.cc
+++ b/src/ic/ic.cc
@@ -3685,11 +3685,10 @@ RUNTIME_FUNCTION(Runtime_CloneObjectIC_Miss) {
             if (CanFastCloneObjectToObjectLiteral(
                     source_map, result_map, {}, null_proto_literal, isolate)) {
               DCHECK(result_map->OnlyHasSimpleProperties());
-              DCHECK_LE(source_map->GetInObjectProperties() -
+              DCHECK_EQ(source_map->GetInObjectProperties() -
                             source_map->UnusedInObjectProperties(),
-                        result_map->GetInObjectProperties());
-              DCHECK_GE(source_map->GetInObjectProperties(),
-                        result_map->GetInObjectProperties());
+                        result_map->GetInObjectProperties() -
+                            result_map->UnusedInObjectProperties());
               UpdateState(result_map);
             } else {
               if (CanCacheCloneTargetMapTransition(
diff --git a/test/mjsunit/clone-ic-regressions.js b/test/mjsunit/clone-ic-regressions.js
index dd916fbe8a2..8217376df4c 100644
--- a/test/mjsunit/clone-ic-regressions.js
+++ b/test/mjsunit/clone-ic-regressions.js
@@ -325,3 +325,18 @@ test({0:0,1:0}, idView);
       ...v7,
   };
 })();
+
+// A case where we copy from a smaller into a bigger object.
+(function() {
+  function F0() {
+  }
+  function F3() {
+      const v9 = new F0();
+      const o10 = {
+          ...v9,
+      };
+  }
+  new F3();
+  new F3();
+  new F3();
+})();
