diff --git a/src/compiler/turbofan-typer.cc b/src/compiler/turbofan-typer.cc
index 9c7c744cb38..62d4a3aba5d 100644
--- a/src/compiler/turbofan-typer.cc
+++ b/src/compiler/turbofan-typer.cc
@@ -722,7 +722,7 @@ Type Typer::Visitor::ObjectIsArrayBufferView(Type type, Typer* t) {
   // TODO(turbofan): Introduce a Type::ArrayBufferView?
   CHECK(!type.IsNone());
   if (type.Is(Type::TypedArray())) return t->singleton_true_;
-  if (!type.Maybe(Type::OtherObject())) return t->singleton_false_;
+  if (!type.Maybe(Type::TypedArray())) return t->singleton_false_;
   return Type::Boolean();
 }
 
diff --git a/test/mjsunit/compiler/regress-398431403.js b/test/mjsunit/compiler/regress-398431403.js
new file mode 100644
index 00000000000..9a21fb1d378
--- /dev/null
+++ b/test/mjsunit/compiler/regress-398431403.js
@@ -0,0 +1,23 @@
+// Copyright 2025 the V8 project authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+// Flags: --allow-natives-syntax --turbofan --no-always-turbofan
+
+const v1 = new BigInt64Array();
+
+function f2(acc, current) {
+  return ArrayBuffer.isView(acc);
+}
+%PrepareFunctionForOptimization(f2);
+
+function f6() {
+  let arr = [{}];
+  return (arr).reduceRight(f2, v1);
+}
+%PrepareFunctionForOptimization(f6);
+
+f6();
+f6();
+%OptimizeFunctionOnNextCall(f6);
+f6();
