diff --git a/src/objects/elements.cc b/src/objects/elements.cc
index 4189d7315a5..15814a25ea1 100644
--- a/src/objects/elements.cc
+++ b/src/objects/elements.cc
@@ -2316,6 +2316,7 @@ class FastElementsAccessor : public ElementsAccessorBase<Subclass, KindTraits> {
     Tagged<BackingStore> backing_store = Cast<BackingStore>(elements);
     DCHECK(length <= std::numeric_limits<int>::max());
     int length_int = static_cast<int>(length);
+    int capacity_int = static_cast<int>(backing_store->length());
     if (IsSmiElementsKind(KindTraits::Kind)) {
       HandleScope scope(isolate);
       for (int i = 0; i < length_int; i++) {
@@ -2331,6 +2332,12 @@ class FastElementsAccessor : public ElementsAccessorBase<Subclass, KindTraits> {
     } else {
       DCHECK(IsHoleyElementsKind(KindTraits::Kind));
     }
+    // Any values in the backing store outside of the length have to be holes,
+    // even if it is PACKED, in case it is extended (and made HOLEY as part of
+    // that extension).
+    for (int i = length_int; i < capacity_int; i++) {
+      DCHECK(backing_store->is_the_hole(isolate, i));
+    }
 #endif
 #endif
   }
diff --git a/src/wasm/wasm-module.cc b/src/wasm/wasm-module.cc
index e21497a97f1..f4b0e5210ff 100644
--- a/src/wasm/wasm-module.cc
+++ b/src/wasm/wasm-module.cc
@@ -513,6 +513,11 @@ DirectHandle<JSArray> GetImports(Isolate* isolate,
   }
 
   array_object->set_length(Smi::FromInt(cursor));
+  // Make sure that values after the cursor are holes.
+  for (int index = cursor; index < num_imports; ++index) {
+    storage->set_the_hole(isolate, cursor++);
+  }
+  JSObject::ValidateElements(*array_object);
   return array_object;
 }
 
@@ -538,7 +543,7 @@ DirectHandle<JSArray> GetExports(Isolate* isolate,
       factory->NewJSArray(PACKED_ELEMENTS, 0, 0);
   DirectHandle<FixedArray> storage = factory->NewFixedArray(num_exports);
   JSArray::SetContent(array_object, storage);
-  array_object->set_length(Smi::FromInt(num_exports));
+  DCHECK_EQ(array_object->length(), Smi::FromInt(num_exports));
 
   DirectHandle<JSFunction> object_function = DirectHandle<JSFunction>(
       isolate->native_context()->object_function(), isolate);
@@ -610,6 +615,7 @@ DirectHandle<JSArray> GetExports(Isolate* isolate,
     storage->set(index, *entry);
   }
 
+  JSObject::ValidateElements(*array_object);
   return array_object;
 }
 
@@ -656,12 +662,13 @@ DirectHandle<JSArray> GetCustomSections(
   DirectHandle<FixedArray> storage =
       factory->NewFixedArray(num_custom_sections);
   JSArray::SetContent(array_object, storage);
-  array_object->set_length(Smi::FromInt(num_custom_sections));
+  DCHECK_EQ(array_object->length(), Smi::FromInt(num_custom_sections));
 
   for (int i = 0; i < num_custom_sections; i++) {
     storage->set(i, *matching_sections[i]);
   }
 
+  JSObject::ValidateElements(*array_object);
   return array_object;
 }
 
diff --git a/test/mjsunit/regress/wasm/regress-crbug-408253898.js b/test/mjsunit/regress/wasm/regress-crbug-408253898.js
new file mode 100644
index 00000000000..00ce4c8366c
--- /dev/null
+++ b/test/mjsunit/regress/wasm/regress-crbug-408253898.js
@@ -0,0 +1,25 @@
+// Copyright 2025 the V8 project authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+// Flags: --allow-natives-syntax
+
+d8.file.execute('test/mjsunit/wasm/wasm-module-builder.js');
+
+let module_builder = new WasmModuleBuilder();
+module_builder.addImportedGlobal("foo", "oh", kWasmExternRef);
+module_builder.addImportedGlobal("bar", "no", kWasmExternRef);
+let wasm_imports = WebAssembly.Module.imports(module_builder.toModule({
+  importedStringConstants: "foo"
+}));
+
+assertEquals(wasm_imports.length, 1);
+assertTrue(%HasFastPackedElements(wasm_imports));
+assertTrue(%HeapObjectVerify(wasm_imports));
+
+wasm_imports.fill(0);
+
+assertEquals(wasm_imports.length, 1);
+assertTrue(%HasSmiElements(wasm_imports));
+assertTrue(%HasFastPackedElements(wasm_imports));
+assertTrue(%HeapObjectVerify(wasm_imports));
