diff --git a/src/objects/waiter-queue-node.cc b/src/objects/waiter-queue-node.cc
index fcd2ba4b5b2..1e0306d4828 100644
--- a/src/objects/waiter-queue-node.cc
+++ b/src/objects/waiter-queue-node.cc
@@ -18,14 +18,14 @@ WaiterQueueNode::~WaiterQueueNode() {
   // Since waiter queue nodes are allocated on the stack, they must be removed
   // from the intrusive linked list once they go out of scope, otherwise there
   // will be dangling pointers.
-  VerifyNotInList();
+  SbxCheckNotInList();
 }
 
 // static
 void WaiterQueueNode::Enqueue(WaiterQueueNode** head,
                               WaiterQueueNode* new_tail) {
   DCHECK_NOT_NULL(head);
-  new_tail->VerifyNotInList();
+  new_tail->SbxCheckNotInList();
   WaiterQueueNode* current_head = *head;
   if (current_head == nullptr) {
     new_tail->next_ = new_tail;
@@ -43,6 +43,7 @@ void WaiterQueueNode::Enqueue(WaiterQueueNode** head,
 void WaiterQueueNode::DequeueUnchecked(WaiterQueueNode** head) {
   if (next_ == this) {
     // The queue contains exactly 1 node.
+    DCHECK_EQ(this, *head);
     *head = nullptr;
   } else {
     // The queue contains >1 nodes.
@@ -156,15 +157,13 @@ uint32_t WaiterQueueNode::NotifyAllInList() {
   return count;
 }
 
-void WaiterQueueNode::VerifyNotInList() {
-  DCHECK_NULL(next_);
-  DCHECK_NULL(prev_);
+void WaiterQueueNode::SbxCheckNotInList() const {
+  SBXCHECK_EQ(nullptr, next_);
+  SBXCHECK_EQ(nullptr, prev_);
 }
 
 void WaiterQueueNode::SetNotInListForVerification() {
-#ifdef DEBUG
   next_ = prev_ = nullptr;
-#endif
 }
 
 }  // namespace detail
diff --git a/src/objects/waiter-queue-node.h b/src/objects/waiter-queue-node.h
index 3028a276823..9ac4e4263af 100644
--- a/src/objects/waiter-queue-node.h
+++ b/src/objects/waiter-queue-node.h
@@ -31,7 +31,7 @@ namespace detail {
 // When compressing pointers (including when sandboxing), the access to the
 // node is indirected through the shared external pointer table.
 //
-// The WaiterQueueNode is an abstract class encapsulting the general queue
+// The WaiterQueueNode is an abstract class encapsulating the general queue
 // logic (enqueue, dequeue, etc...). Its extensions add the logic to handle
 // notifications and sync/async waiting.
 // TODO(v8:12547): Unittest this.
@@ -87,7 +87,7 @@ class V8_NODISCARD WaiterQueueNode {
 
  private:
   void DequeueUnchecked(WaiterQueueNode** head);
-  void VerifyNotInList();
+  void SbxCheckNotInList() const;
 };
 
 }  // namespace detail
diff --git a/test/mjsunit/sandbox/regress-443182220.js b/test/mjsunit/sandbox/regress-443182220.js
index 9e0826a23a7..749cc685a43 100644
--- a/test/mjsunit/sandbox/regress-443182220.js
+++ b/test/mjsunit/sandbox/regress-443182220.js
@@ -7,41 +7,34 @@
 function findForeign(start_addr, max_objects = 35) {
   const tasks = [];
   let addr = start_addr;
-  var foreign_count = 0;
   for (let i = 0; i < max_objects; i++) {
     let size = Sandbox.getSizeOfObjectAt(addr);
     objtype = Sandbox.getInstanceTypeOfObjectAt(addr)
-    print(objtype + " at 0x" + (addr+1+Sandbox.base).toString(16));
-    if (objtype == "PROMISE_REACTION_TYPE"){
-      print("====================================")
-      reject_handler = r32(addr+12);
-      print("reject_handler: 0x" + (reject_handler+Sandbox.base).toString(16));
-      context = r32(reject_handler-1+20);
-      print("context: 0x" + (context+Sandbox.base).toString(16));
-      foreign = r32(context-1+24);
-      print("foreign: 0x" + (foreign+Sandbox.base).toString(16));
-      tasks.push(context-1);
-      print("====================================")
+    if (objtype == "PROMISE_REACTION_TYPE") {
+      reject_handler = r32(addr + 12);
+      context = r32(reject_handler - 1 + 20);
+      foreign = r32(context - 1 + 24);
+      tasks.push(context - 1);
     }
     addr = (addr + size) >>> 0;
   }
   return tasks;
 }
+
 const mem = new DataView(new Sandbox.MemoryView(0, 0x100000000));
 
 const r32 = (addr) => mem.getUint32(addr, true);
 const w32 = (addr, val) => mem.setUint32(addr, val, true);
 
-oo={};
+oo = {};
 (async () => {
-  o = {};
   const m1 = new Atomics.Mutex();
   const m2 = new Atomics.Mutex();
   const p1 = Atomics.Mutex.lockAsync(m1, () => 1);
   const p2 = Atomics.Mutex.lockAsync(m2, () => 2);
 
   tasks = findForeign(Sandbox.getAddressOf(oo));
-  w32(tasks[1]+24, r32(tasks[0]+24));
+  w32(tasks[1] + 24, r32(tasks[0] + 24));
 
   await Promise.allSettled([p1, p2]);
 })();
diff --git a/test/mjsunit/sandbox/regress-447307165.js b/test/mjsunit/sandbox/regress-447307165.js
new file mode 100644
index 00000000000..a1d91f13afb
--- /dev/null
+++ b/test/mjsunit/sandbox/regress-447307165.js
@@ -0,0 +1,39 @@
+// Copyright 2025 the V8 project authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+// Flags: --harmony-struct --sandbox-testing
+
+function findForeign(start_addr, max_objects = 35) {
+  const tasks = [];
+  let addr = start_addr;
+  for (let i = 0; i < max_objects; i++) {
+    let size = Sandbox.getSizeOfObjectAt(addr);
+    objtype = Sandbox.getInstanceTypeOfObjectAt(addr)
+    if (objtype == "PROMISE_REACTION_TYPE") {
+      reject_handler = r32(addr + 12);
+      context = r32(reject_handler - 1 + 20);
+      foreign = r32(context - 1 + 24);
+      tasks.push(context - 1);
+    }
+    addr = (addr + size) >>> 0;
+  }
+  return tasks;
+}
+
+const mem = new DataView(new Sandbox.MemoryView(0, 0x100000000));
+
+const r32 = (addr) => mem.getUint32(addr, true);
+const w32 = (addr, val) => mem.setUint32(addr, val, true);
+
+oo = {};
+(async () => {
+  const m = new Atomics.Mutex();
+  const p1 = Atomics.Mutex.lockAsync(m, () => 1);
+  const p2 = Atomics.Mutex.lockAsync(m, () => 2);
+
+  tasks = findForeign(Sandbox.getAddressOf(oo));
+  w32(tasks[0] + 24, r32(tasks[2] + 24));
+
+  await Promise.allSettled([p1, p2]);
+})();
