diff --git a/src/compiler/access-info.cc b/src/compiler/access-info.cc
index 016c0297476..bcce591efe0 100644
--- a/src/compiler/access-info.cc
+++ b/src/compiler/access-info.cc
@@ -1168,18 +1168,20 @@ PropertyAccessInfo AccessInfoFactory::LookupSpecialFieldAccessorInHolder(
     Tagged<Object> maybe_accessors = descriptors->GetStrongValue(index);
     if (IsAccessorPair(maybe_accessors)) {
       Tagged<AccessorPair> accessors = Cast<AccessorPair>(maybe_accessors);
-      Tagged<JSFunction> getter =
-          Cast<JSFunction>(accessors->getter(kAcquireLoad));
-      if (getter->shared()->HasBuiltinId() &&
-          getter->shared()->builtin_id() ==
-              Builtin::kTypedArrayPrototypeLength &&
-          broker_->dependencies()->DependOnArrayBufferDetachingProtector()) {
-        dependencies()->DependOnStablePrototypeChain(receiver_map,
-                                                     kStartAtPrototype, holder);
-        // TODO(388844115): If we cannot depend on the detaching protector, add
-        // a different kind of TypedArrayLength operator which checks for
-        // detached before reading the byte_length.
-        return PropertyAccessInfo::TypedArrayLength(zone(), receiver_map);
+      Tagged<Object> maybe_getter = accessors->getter(kAcquireLoad);
+      if (Tagged<JSFunction> getter;
+          TryCast<JSFunction>(maybe_getter, &getter)) {
+        if (getter->shared()->HasBuiltinId() &&
+            getter->shared()->builtin_id() ==
+                Builtin::kTypedArrayPrototypeLength &&
+            broker_->dependencies()->DependOnArrayBufferDetachingProtector()) {
+          dependencies()->DependOnStablePrototypeChain(
+              receiver_map, kStartAtPrototype, holder);
+          // TODO(388844115): If we cannot depend on the detaching protector,
+          // add a different kind of TypedArrayLength operator which checks for
+          // detached before reading the byte_length.
+          return PropertyAccessInfo::TypedArrayLength(zone(), receiver_map);
+        }
       }
     }
   }
diff --git a/test/mjsunit/compiler/regress-419744895.js b/test/mjsunit/compiler/regress-419744895.js
new file mode 100644
index 00000000000..6b749c1eccb
--- /dev/null
+++ b/test/mjsunit/compiler/regress-419744895.js
@@ -0,0 +1,15 @@
+// Copyright 2025 the V8 project authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+// Flags: --allow-natives-syntax --turbofan --no-always-turbofan
+
+Object.defineProperty(Uint8Array.prototype, 'length', { get: undefined });
+function foo() {
+  let v = new Uint8Array();
+  return v.length;
+}
+%PrepareFunctionForOptimization(foo);
+foo();
+%OptimizeFunctionOnNextCall(foo);
+foo();
diff --git a/test/mjsunit/maglev/regress-419744895.js b/test/mjsunit/maglev/regress-419744895.js
new file mode 100644
index 00000000000..eff5cc62c95
--- /dev/null
+++ b/test/mjsunit/maglev/regress-419744895.js
@@ -0,0 +1,15 @@
+// Copyright 2025 the V8 project authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+// Flags: --allow-natives-syntax --maglev --no-always-turbofan
+
+Object.defineProperty(Uint8Array.prototype, 'length', { get: undefined });
+function foo() {
+  let v = new Uint8Array();
+  return v.length;
+}
+%PrepareFunctionForOptimization(foo);
+foo();
+%OptimizeMaglevOnNextCall(foo);
+foo();
