diff --git a/src/ic/binary-op-assembler.cc b/src/ic/binary-op-assembler.cc
index d5901c9996b..e9b6b9446eb 100644
--- a/src/ic/binary-op-assembler.cc
+++ b/src/ic/binary-op-assembler.cc
@@ -107,7 +107,9 @@ TNode<Object> BinaryOpAssembler::Generate_AddWithFeedback(
         var_fadd_lhs = SmiToFloat64(lhs_smi);
         var_fadd_rhs = SmiToFloat64(rhs_smi);
         var_type_feedback =
-            SmiConstant(BinaryOperationFeedback::kAdditiveSafeInteger);
+            SelectSmiConstant(IsAdditiveSafeIntegerFeedbackEnabled(),
+                              BinaryOperationFeedback::kAdditiveSafeInteger,
+                              BinaryOperationFeedback::kNumber);
         Goto(&do_fadd);
       }
     }
diff --git a/test/mjsunit/regress-396485545.js b/test/mjsunit/regress-396485545.js
new file mode 100644
index 00000000000..d91dc5f6400
--- /dev/null
+++ b/test/mjsunit/regress-396485545.js
@@ -0,0 +1,16 @@
+// Copyright 2025 the V8 project authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+// Flags: --allow-natives-syntax
+
+function foo(a) {
+  a += 1;
+  a += 1;
+  return a;
+}
+
+%PrepareFunctionForOptimization(foo);
+assertEquals(1073741825, foo(1073741823));
+%OptimizeFunctionOnNextCall(foo);
+assertEquals(NaN, foo());
diff --git a/test/unittests/interpreter/interpreter-unittest.cc b/test/unittests/interpreter/interpreter-unittest.cc
index dc04ae7ceb3..309a8bff968 100644
--- a/test/unittests/interpreter/interpreter-unittest.cc
+++ b/test/unittests/interpreter/interpreter-unittest.cc
@@ -606,7 +606,9 @@ TEST_F(InterpreterTest, InterpreterBinaryOpTypeFeedback) {
        BinaryOperationFeedback::kSignedSmall},
       {Token::kAdd, LiteralForTest(Smi::kMaxValue), LiteralForTest(1),
        i_isolate()->factory()->NewHeapNumber(Smi::kMaxValue + 1.0),
-       BinaryOperationFeedback::kAdditiveSafeInteger},
+       v8_flags.additive_safe_int_feedback
+           ? BinaryOperationFeedback::kAdditiveSafeInteger
+           : BinaryOperationFeedback::kNumber},
       {Token::kAdd, LiteralForTest(3.1415), LiteralForTest(3),
        i_isolate()->factory()->NewHeapNumber(3.1415 + 3),
        BinaryOperationFeedback::kNumber},
@@ -740,7 +742,9 @@ TEST_F(InterpreterTest, InterpreterBinaryOpSmiTypeFeedback) {
        BinaryOperationFeedback::kSignedSmall},
       {Token::kAdd, LiteralForTest(2), Smi::kMaxValue,
        i_isolate()->factory()->NewHeapNumber(Smi::kMaxValue + 2.0),
-       BinaryOperationFeedback::kAdditiveSafeInteger},
+       v8_flags.additive_safe_int_feedback
+           ? BinaryOperationFeedback::kAdditiveSafeInteger
+           : BinaryOperationFeedback::kNumber},
       {Token::kAdd, LiteralForTest(3.1415), 2,
        i_isolate()->factory()->NewHeapNumber(3.1415 + 2.0),
        BinaryOperationFeedback::kNumber},
