diff --git a/src/deoptimizer/translated-state.cc b/src/deoptimizer/translated-state.cc
index a2b5fc2a2e0..c82ed3495f3 100644
--- a/src/deoptimizer/translated-state.cc
+++ b/src/deoptimizer/translated-state.cc
@@ -2458,9 +2458,16 @@ void TranslatedState::InitializeJSObjectAt(
     } else if (InstanceTypeChecker::IsJSRegExp(map->instance_type()) &&
                offset == JSRegExp::kDataOffset) {
       DirectHandle<HeapObject> field_value = slot->storage();
-      CHECK(IsRegExpDataWrapper(*field_value));
-      Tagged<RegExpData> value =
-          Cast<RegExpDataWrapper>(*field_value)->data(isolate());
+      // If the value comes from the DeoptimizationLiteralArray, it is a
+      // RegExpDataWrapper as we can't store TrustedSpace values in a FixedArray
+      // directly.
+      Tagged<RegExpData> value;
+      if (Is<RegExpDataWrapper>(*field_value)) {
+        value = Cast<RegExpDataWrapper>(*field_value)->data(isolate());
+      } else {
+        CHECK(IsRegExpData(*field_value));
+        value = Cast<RegExpData>(*field_value);
+      }
       object_storage
           ->RawIndirectPointerField(offset, kRegExpDataIndirectPointerTag)
           .Relaxed_Store(value);
diff --git a/test/mjsunit/regress/regress-359618508.js b/test/mjsunit/regress/regress-359618508.js
new file mode 100644
index 00000000000..822179c5896
--- /dev/null
+++ b/test/mjsunit/regress/regress-359618508.js
@@ -0,0 +1,28 @@
+// Copyright 2024 the V8 project authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+//
+// Flags: --allow-natives-syntax
+
+function foo(a3,  a6) {
+  a3.f = a6;
+  const v8 = new Int32Array();
+  v8[148] = v8;
+}
+
+function bar() {
+  return foo(/xvxyz{1,32}?(ab|cde)\1eFa\S/suimyg);
+}
+
+class C0 {
+}
+const v1 = new C0();
+
+%PrepareFunctionForOptimization(foo);
+%PrepareFunctionForOptimization(bar);
+foo(v1);
+
+for (let v13 = 0; v13 < 5; v13++) {
+  bar();
+  %OptimizeFunctionOnNextCall(bar);
+}
