diff --git a/src/init/isolate-group.cc b/src/init/isolate-group.cc
index d0843d71bfb..9af2277d514 100644
--- a/src/init/isolate-group.cc
+++ b/src/init/isolate-group.cc
@@ -589,6 +589,10 @@ class PABackedSandboxedArrayBufferAllocator::Impl final {
     }
     // The V8 sandbox is guaranteed to be large enough to host the pool.
     CHECK(pool_base);
+    // Call PartitionAddressSpace::Init() first just to make sure metadata
+    // region start is initialized and the configurable pool allocations do have
+    // out-of-line metadata.
+    partition_alloc::internal::PartitionAddressSpace::Init();
     partition_alloc::internal::PartitionAddressSpace::InitConfigurablePool(
         pool_base, pool_size);
 
