diff --git a/src/maglev/maglev-graph-builder.cc b/src/maglev/maglev-graph-builder.cc
index e212f930edf..0046b794d95 100644
--- a/src/maglev/maglev-graph-builder.cc
+++ b/src/maglev/maglev-graph-builder.cc
@@ -13029,7 +13029,13 @@ ValueNode* MaglevGraphBuilder::BuildInlinedAllocation(
                   CreateHeapNumber(node->Cast<Float64Constant>()->value()),
                   allocation_type);
             } else {
-              node = GetTaggedValue(node);
+              ValueNode* new_node = GetTaggedValue(node);
+              if (new_node != node && new_node->properties().can_allocate()) {
+                // TODO(olivf): Remove this and instead always clear when we
+                // emit an allocating instruction.
+                ClearCurrentAllocationBlock();
+              }
+              node = new_node;
             }
             values.push_back(node);
           });
