diff --git a/src/ast/scopes.cc b/src/ast/scopes.cc
index 76d291e5bc2..3f0e8b3a393 100644
--- a/src/ast/scopes.cc
+++ b/src/ast/scopes.cc
@@ -2705,6 +2705,7 @@ void Scope::AllocateScopeInfosRecursively(
     // next outer scope that needs a context.
     next_outer_scope = scope_info_;
     DCHECK(!scope_info_.is_null());
+    DCHECK(!is_hidden());
     CHECK_EQ(scope_info_->scope_type(), scope_type_);
     CHECK_EQ(scope_info_->ContextLength(), num_heap_slots_);
 #ifdef DEBUG
@@ -2713,8 +2714,9 @@ void Scope::AllocateScopeInfosRecursively(
 #endif
   } else if (NeedsScopeInfo()) {
 #ifdef DEBUG
-    // Mark this ID as being used.
-    scope_infos_to_reuse[UniqueIdInScript()] = {};
+    // Mark this ID as being used. Skip hidden scopes because they are
+    // synthetic, unreusable, but hard to make unique.
+    if (!is_hidden()) scope_infos_to_reuse[UniqueIdInScript()] = {};
 #endif
     scope_info_ = ScopeInfo::Create(isolate, zone(), this, outer_scope);
     DCHECK_EQ(UniqueIdInScript(), scope_info_->UniqueIdInScript());
diff --git a/test/mjsunit/regress-4530868594868224.js b/test/mjsunit/regress-4530868594868224.js
new file mode 100644
index 00000000000..ddd1363a9a1
--- /dev/null
+++ b/test/mjsunit/regress-4530868594868224.js
@@ -0,0 +1,22 @@
+// Copyright 2024 the V8 project authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+// Flags: --fuzzing
+
+const __v_2 = [{
+  body: function () {
+  },
+  body: function () {
+  }
+}, {
+  body: function () {
+    const __v_11 = {
+    };
+  }
+}];
+  var __v_30 = [{
+    body: function () {
+        var __v_140 = async () => __v_147 = async () => __v_173 = 'str';
+    }
+  }];
