diff --git a/src/wasm/wasm-js.cc b/src/wasm/wasm-js.cc
index 0a84b57dd9a..d42a3dd2f72 100644
--- a/src/wasm/wasm-js.cc
+++ b/src/wasm/wasm-js.cc
@@ -3584,45 +3584,57 @@ bool WasmJs::InstallTypeReflection(Isolate* isolate,
     return false;
   }
 
-  Handle<String> type_string = v8_str(isolate, "type");
-#define INSTANCE_PROTO_HANDLE(Name) \
-  handle(Cast<JSObject>(context->Name()->instance_prototype()), isolate)
+  Handle<JSObject> table_proto(
+      Cast<JSObject>(context->wasm_table_constructor()->instance_prototype()),
+      isolate);
+  Handle<JSObject> global_proto(
+      Cast<JSObject>(context->wasm_global_constructor()->instance_prototype()),
+      isolate);
+  Handle<JSObject> memory_proto(
+      Cast<JSObject>(context->wasm_memory_constructor()->instance_prototype()),
+      isolate);
+  Handle<JSObject> tag_proto(
+      Cast<JSObject>(context->wasm_tag_constructor()->instance_prototype()),
+      isolate);
 
-  if (JSObject::HasRealNamedProperty(
-          isolate, INSTANCE_PROTO_HANDLE(wasm_table_constructor), type_string)
+  Handle<String> type_string = v8_str(isolate, "type");
+  if (JSObject::HasRealNamedProperty(isolate, table_proto, type_string)
           .FromMaybe(true)) {
     return false;
   }
-  if (JSObject::HasRealNamedProperty(
-          isolate, INSTANCE_PROTO_HANDLE(wasm_global_constructor), type_string)
+  if (JSObject::HasRealNamedProperty(isolate, global_proto, type_string)
           .FromMaybe(true)) {
     return false;
   }
-  if (JSObject::HasRealNamedProperty(
-          isolate, INSTANCE_PROTO_HANDLE(wasm_memory_constructor), type_string)
+  if (JSObject::HasRealNamedProperty(isolate, memory_proto, type_string)
           .FromMaybe(true)) {
     return false;
   }
-  if (JSObject::HasRealNamedProperty(
-          isolate, INSTANCE_PROTO_HANDLE(wasm_tag_constructor), type_string)
+  if (JSObject::HasRealNamedProperty(isolate, tag_proto, type_string)
           .FromMaybe(true)) {
     return false;
   }
 
+  // Ensure prototype objects are extensible, otherwise adding properties
+  // to them will fail. Extensibility of the `WebAssembly` object should
+  // already have been checked by the caller.
+  DCHECK(webassembly->map()->is_extensible());
+  if (!table_proto->map()->is_extensible() ||
+      !global_proto->map()->is_extensible() ||
+      !memory_proto->map()->is_extensible() ||
+      !tag_proto->map()->is_extensible()) {
+    return false;
+  }
+
   // Checks are done, start installing the new fields.
-  InstallFunc(isolate, INSTANCE_PROTO_HANDLE(wasm_table_constructor), "type",
-              WebAssemblyTableType, 0, false, NONE,
-              SideEffectType::kHasNoSideEffect);
-  InstallFunc(isolate, INSTANCE_PROTO_HANDLE(wasm_memory_constructor), "type",
-              WebAssemblyMemoryType, 0, false, NONE,
-              SideEffectType::kHasNoSideEffect);
-  InstallFunc(isolate, INSTANCE_PROTO_HANDLE(wasm_global_constructor), "type",
-              WebAssemblyGlobalType, 0, false, NONE,
-              SideEffectType::kHasNoSideEffect);
-  InstallFunc(isolate, INSTANCE_PROTO_HANDLE(wasm_tag_constructor), "type",
-              WebAssemblyTagType, 0, false, NONE,
+  InstallFunc(isolate, table_proto, "type", WebAssemblyTableType, 0, false,
+              NONE, SideEffectType::kHasNoSideEffect);
+  InstallFunc(isolate, memory_proto, "type", WebAssemblyMemoryType, 0, false,
+              NONE, SideEffectType::kHasNoSideEffect);
+  InstallFunc(isolate, global_proto, "type", WebAssemblyGlobalType, 0, false,
+              NONE, SideEffectType::kHasNoSideEffect);
+  InstallFunc(isolate, tag_proto, "type", WebAssemblyTagType, 0, false, NONE,
               SideEffectType::kHasNoSideEffect);
-#undef INSTANCE_PROTO_HANDLE
 
   // Create the Function object.
   Handle<JSFunction> function_constructor = InstallConstructorFunc(
diff --git a/test/mjsunit/regress/wasm/regress-348473836-global.js b/test/mjsunit/regress/wasm/regress-348473836-global.js
new file mode 100644
index 00000000000..b5d6767c469
--- /dev/null
+++ b/test/mjsunit/regress/wasm/regress-348473836-global.js
@@ -0,0 +1,6 @@
+// Copyright 2024 the V8 project authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+Object.seal(WebAssembly.Global.prototype);
+d8.test.enableJSPI();
diff --git a/test/mjsunit/regress/wasm/regress-348473836-memory.js b/test/mjsunit/regress/wasm/regress-348473836-memory.js
new file mode 100644
index 00000000000..167b6b115f4
--- /dev/null
+++ b/test/mjsunit/regress/wasm/regress-348473836-memory.js
@@ -0,0 +1,6 @@
+// Copyright 2024 the V8 project authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+Object.seal(WebAssembly.Memory.prototype);
+d8.test.enableJSPI();
diff --git a/test/mjsunit/regress/wasm/regress-348473836-table.js b/test/mjsunit/regress/wasm/regress-348473836-table.js
new file mode 100644
index 00000000000..0057459f1b4
--- /dev/null
+++ b/test/mjsunit/regress/wasm/regress-348473836-table.js
@@ -0,0 +1,6 @@
+// Copyright 2024 the V8 project authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+Object.seal(WebAssembly.Table.prototype);
+d8.test.enableJSPI();
diff --git a/test/mjsunit/regress/wasm/regress-348473836-tag.js b/test/mjsunit/regress/wasm/regress-348473836-tag.js
new file mode 100644
index 00000000000..0e878617053
--- /dev/null
+++ b/test/mjsunit/regress/wasm/regress-348473836-tag.js
@@ -0,0 +1,6 @@
+// Copyright 2024 the V8 project authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+Object.seal(WebAssembly.Tag.prototype);
+d8.test.enableJSPI();
diff --git a/test/mjsunit/regress/wasm/regress-348473836-webassembly.js b/test/mjsunit/regress/wasm/regress-348473836-webassembly.js
new file mode 100644
index 00000000000..f6940b53948
--- /dev/null
+++ b/test/mjsunit/regress/wasm/regress-348473836-webassembly.js
@@ -0,0 +1,6 @@
+// Copyright 2024 the V8 project authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+Object.seal(WebAssembly);
+d8.test.enableJSPI();
