diff --git a/src/maglev/maglev-reducer-inl.h b/src/maglev/maglev-reducer-inl.h
index f40ed29d0f5..2981af658a8 100644
--- a/src/maglev/maglev-reducer-inl.h
+++ b/src/maglev/maglev-reducer-inl.h
@@ -1563,7 +1563,7 @@ MaybeReduceResult MaglevReducer<BaseT>::TryFoldFloat64Min(ValueNode* lhs,
   DCHECK(ValueRepresentationIs(rhs->value_representation(),
                                ValueRepresentation::kHoleyFloat64));
   if (lhs == rhs) {
-    return lhs;
+    return lhs->Unwrap();
   }
 
   std::optional<double> lhs_const =
@@ -1594,6 +1594,16 @@ MaybeReduceResult MaglevReducer<BaseT>::TryFoldFloat64Min(ValueNode* lhs,
 template <typename BaseT>
 MaybeReduceResult MaglevReducer<BaseT>::TryFoldFloat64Max(ValueNode* lhs,
                                                           ValueNode* rhs) {
+  // lhs and rhs need to be already converted to HoleyFloat64. Otherwise
+  // equality checking is not valid.
+  DCHECK(ValueRepresentationIs(lhs->value_representation(),
+                               ValueRepresentation::kHoleyFloat64));
+  DCHECK(ValueRepresentationIs(rhs->value_representation(),
+                               ValueRepresentation::kHoleyFloat64));
+  if (lhs == rhs) {
+    return lhs->Unwrap();
+  }
+
   std::optional<double> lhs_const =
       TryGetFloat64Constant(UseRepresentation::kFloat64, lhs,
                             TaggedToFloat64ConversionType::kNumberOrOddball);
diff --git a/test/mjsunit/maglev/regress-454364323.js b/test/mjsunit/maglev/regress-454364323.js
new file mode 100644
index 00000000000..1a5e74d3c7a
--- /dev/null
+++ b/test/mjsunit/maglev/regress-454364323.js
@@ -0,0 +1,14 @@
+// Copyright 2025 the V8 project authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+// Flags: --allow-natives-syntax --maglev
+
+function foo(v1) {
+  let v2 = v1 >>> 0;
+  return Math.min(v2, v2);
+}
+%PrepareFunctionForOptimization(foo);
+foo();
+%OptimizeMaglevOnNextCall(foo);
+foo();
