diff --git a/src/common/ptr-compr-inl.h b/src/common/ptr-compr-inl.h
index 742d1daf31b..f27ae78dfc1 100644
--- a/src/common/ptr-compr-inl.h
+++ b/src/common/ptr-compr-inl.h
@@ -85,7 +85,8 @@ Tagged_t V8HeapCompressionSchemeImpl<Cage>::CompressObject(Address tagged) {
   // This is used to help clang produce better code. Values which could be
   // invalid pointers need to be compressed with CompressAny.
 #ifdef V8_COMPRESS_POINTERS_IN_SHARED_CAGE
-  V8_ASSUME((tagged & kPtrComprCageBaseMask) == base() || HAS_SMI_TAG(tagged));
+  DCHECK_IMPLIES(!HAS_SMI_TAG(tagged),
+                 (tagged & kPtrComprCageBaseMask) == base());
 #endif
   return static_cast<Tagged_t>(tagged);
 }
diff --git a/src/common/ptr-compr.h b/src/common/ptr-compr.h
index f5a854bbe0e..cdc9cd81452 100644
--- a/src/common/ptr-compr.h
+++ b/src/common/ptr-compr.h
@@ -104,7 +104,7 @@ class SmiCompressionScheme : public AllStatic {
   }
 
   static Tagged_t CompressObject(Address tagged) {
-    V8_ASSUME(HAS_SMI_TAG(tagged));
+    DCHECK(HAS_SMI_TAG(tagged));
     return static_cast<Tagged_t>(tagged);
   }
 };
diff --git a/src/objects/tagged-field-inl.h b/src/objects/tagged-field-inl.h
index 883d3a86026..091938199dd 100644
--- a/src/objects/tagged-field-inl.h
+++ b/src/objects/tagged-field-inl.h
@@ -21,7 +21,7 @@ Address TaggedMember<T, CompressionScheme>::tagged_to_full(
     Tagged_t tagged_value) {
 #ifdef V8_COMPRESS_POINTERS
   if constexpr (std::is_same_v<Smi, T>) {
-    V8_ASSUME(HAS_SMI_TAG(tagged_value));
+    DCHECK(HAS_SMI_TAG(tagged_value));
     return CompressionScheme::DecompressTaggedSigned(tagged_value);
   } else {
     return CompressionScheme::DecompressTagged(CompressionScheme::base(),
@@ -186,7 +186,7 @@ Address TaggedField<T, kFieldOffset, CompressionScheme>::tagged_to_full(
     TOnHeapAddress on_heap_addr, Tagged_t tagged_value) {
 #ifdef V8_COMPRESS_POINTERS
   if constexpr (kIsSmi) {
-    V8_ASSUME(HAS_SMI_TAG(tagged_value));
+    DCHECK(HAS_SMI_TAG(tagged_value));
     return CompressionScheme::DecompressTaggedSigned(tagged_value);
   } else {
     return CompressionScheme::DecompressTagged(on_heap_addr, tagged_value);
@@ -201,7 +201,7 @@ template <typename T, int kFieldOffset, typename CompressionScheme>
 Tagged_t TaggedField<T, kFieldOffset, CompressionScheme>::full_to_tagged(
     Address value) {
 #ifdef V8_COMPRESS_POINTERS
-  if constexpr (kIsSmi) V8_ASSUME(HAS_SMI_TAG(value));
+  if constexpr (kIsSmi) DCHECK(HAS_SMI_TAG(value));
   return CompressionScheme::CompressObject(value);
 #else
   return value;
