diff --git a/src/wasm/canonical-types.h b/src/wasm/canonical-types.h
index 7c8618fab5a..bcb691372d9 100644
--- a/src/wasm/canonical-types.h
+++ b/src/wasm/canonical-types.h
@@ -325,10 +325,15 @@ class TypeCanonicalizer {
 
     bool EqualStructType(const CanonicalStructType& type1,
                          const CanonicalStructType& type2) const {
-      return std::equal(
-          type1.fields().begin(), type1.fields().end(), type2.fields().begin(),
-          type2.fields().end(),
-          std::bind_front(&CanonicalEquality::EqualValueType, this));
+      return
+          // Compare fields, including a check that the size is the same.
+          std::equal(
+              type1.fields().begin(), type1.fields().end(),
+              type2.fields().begin(), type2.fields().end(),
+              std::bind_front(&CanonicalEquality::EqualValueType, this)) &&
+          // Compare mutabilities, skipping the check for the size.
+          std::equal(type1.mutabilities().begin(), type1.mutabilities().end(),
+                     type2.mutabilities().begin());
     }
 
     bool EqualArrayType(const CanonicalArrayType& type1,
