diff --git a/src/maglev/maglev-graph-builder.cc b/src/maglev/maglev-graph-builder.cc
index f4f61d645b6..b39d0728087 100644
--- a/src/maglev/maglev-graph-builder.cc
+++ b/src/maglev/maglev-graph-builder.cc
@@ -7301,6 +7301,10 @@ MaglevGraphBuilder::FindContinuationForPolymorphicPropertyLoad() {
   // CallProperty REG ...
 
   iterator_.Advance();
+  if (IsOffsetAMergePointOrLoopHeapder(iterator_.current_offset())) {
+    return {};
+  }
+
   switch (iterator_.current_bytecode()) {
 #define CASE(Name, ...)                                                       \
   case interpreter::Bytecode::k##Name:                                        \
@@ -7315,6 +7319,10 @@ MaglevGraphBuilder::FindContinuationForPolymorphicPropertyLoad() {
   }
 
   iterator_.Advance();
+  if (IsOffsetAMergePointOrLoopHeapder(iterator_.current_offset())) {
+    return {};
+  }
+
   switch (iterator_.current_bytecode()) {
 #define CASE(Name, ...)                                                \
   case interpreter::Bytecode::k##Name:                                 \
diff --git a/src/maglev/maglev-graph-builder.h b/src/maglev/maglev-graph-builder.h
index 4b427fd2544..1bb7b2ffa58 100644
--- a/src/maglev/maglev-graph-builder.h
+++ b/src/maglev/maglev-graph-builder.h
@@ -721,6 +721,11 @@ class MaglevGraphBuilder {
     return merge_states_[offset] != nullptr;
   }
 
+  bool IsOffsetAMergePointOrLoopHeapder(int offset) {
+    return IsOffsetAMergePoint(offset) ||
+           bytecode_analysis().IsLoopHeader(offset);
+  }
+
   ValueNode* GetContextAtDepth(ValueNode* context, size_t depth);
   bool CheckContextExtensions(size_t depth);
 
diff --git a/test/mjsunit/maglev/regress-424242619.js b/test/mjsunit/maglev/regress-424242619.js
new file mode 100644
index 00000000000..d873a608c30
--- /dev/null
+++ b/test/mjsunit/maglev/regress-424242619.js
@@ -0,0 +1,19 @@
+// Copyright 2025 the V8 project authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+// Flags: --allow-natives-syntax --no-lazy-feedback-allocation
+// Flags: --maglev-poly-calls
+
+const v1 = [-9223372036854775808125284];
+class C2 {
+  constructor(a4) {
+    const v6 ="MIN_VALUE".toLowerCase(...a4, ...a4);
+    const t4 = this.constructor;
+    new t4(v6);
+  }
+}
+try {
+  new C2(v1);
+} catch {
+}
