diff --git a/src/maglev/maglev-graph-builder.cc b/src/maglev/maglev-graph-builder.cc
index f907ad0651b..18b8b0e899a 100644
--- a/src/maglev/maglev-graph-builder.cc
+++ b/src/maglev/maglev-graph-builder.cc
@@ -1117,8 +1117,8 @@ DeoptFrame MaglevGraphBuilder::GetDeoptFrameForLazyDeoptHelper(
 
   // Currently only support builtin continuations for bytecodes that write to
   // the accumulator
-  DCHECK(
-      interpreter::Bytecodes::WritesAccumulator(iterator_.current_bytecode()));
+  DCHECK(interpreter::Bytecodes::WritesOrClobbersAccumulator(
+      iterator_.current_bytecode()));
 
 #ifdef DEBUG
   if (scope->data().tag() == DeoptFrame::FrameType::kBuiltinContinuationFrame) {
diff --git a/test/mjsunit/maglev/regress-328483400.js b/test/mjsunit/maglev/regress-328483400.js
new file mode 100644
index 00000000000..4f1e33d9abc
--- /dev/null
+++ b/test/mjsunit/maglev/regress-328483400.js
@@ -0,0 +1,17 @@
+// Copyright 2024 the V8 project authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+//
+// Flags: --allow-natives-syntax --maglev
+
+function foo(a) {
+    const o = {};
+    Object.defineProperty(o, "x", { set: Math.ceil });
+    o.x = a; // Creates a Math.ceil builtin continuation.
+}
+
+%PrepareFunctionForOptimization(foo);
+foo({});
+foo({});
+%OptimizeMaglevOnNextCall(foo);
+foo({});
