diff --git a/src/ic/ic.cc b/src/ic/ic.cc
index 6016a682ba9..ab095893300 100644
--- a/src/ic/ic.cc
+++ b/src/ic/ic.cc
@@ -3306,6 +3306,26 @@ FastCloneObjectMode GetCloneModeForMap(Handle<Map> map, bool null_proto_literal,
   return mode;
 }
 
+bool CanCacheCloneTargetMapTransition(Handle<Map> source_map,
+                                      std::optional<Handle<Map>> target_map,
+                                      bool null_proto_literal) {
+  if (!v8_flags.clone_object_sidestep_transitions || null_proto_literal) {
+    return false;
+  }
+  // As of now any R/O source object should end up in the kEmptyObject case, but
+  // there is not really a way of ensuring it. Thus, we also check it below.
+  DCHECK(!InReadOnlySpace(*source_map));
+  if (InReadOnlySpace(*source_map) || source_map->is_deprecated() ||
+      source_map->is_prototype_map()) {
+    return false;
+  }
+  if (!target_map) {
+    return true;
+  }
+  CHECK(!InReadOnlySpace(**target_map));
+  return !(*target_map)->is_deprecated();
+}
+
 bool CanFastCloneObjectWithDifferentMaps(Handle<Map> source_map,
                                          Handle<Map> target_map,
                                          bool null_proto_literal,
@@ -3385,7 +3405,8 @@ bool CanFastCloneObjectWithDifferentMaps(Handle<Map> source_map,
     // In the case the clone also involves a proto transition, we do not cache
     // the target and thus cannot keep track of representation dependencies. We
     // can only allow the most generic target representation.
-    if ((null_proto_literal || !v8_flags.clone_object_sidestep_transitions) &&
+    if (!CanCacheCloneTargetMapTransition(source_map, target_map,
+                                          null_proto_literal) &&
         !details.representation().MostGenericInPlaceChange().Equals(
             target_details.representation())) {
       return false;
@@ -3487,6 +3508,7 @@ std::optional<Tagged<Map>> GetCloneTargetMap(Isolate* isolate,
 void SetCloneTargetMap(Isolate* isolate, Handle<Map> source_map,
                        Handle<Map> new_target_map) {
   if (!v8_flags.clone_object_sidestep_transitions) return;
+  DCHECK(CanCacheCloneTargetMapTransition(source_map, new_target_map, false));
   // Adding this transition also ensures that when the source map field
   // generalizes, we also generalize the target map.
 #ifdef DEBUG
@@ -3507,6 +3529,7 @@ void SetCloneTargetMap(Isolate* isolate, Handle<Map> source_map,
 
 void SetCloneTargetMapUnsupported(Isolate* isolate, Handle<Map> source_map) {
   if (!v8_flags.clone_object_sidestep_transitions) return;
+  DCHECK(CanCacheCloneTargetMapTransition(source_map, {}, false));
   // Adding this transition also ensures that when the source map field
   // generalizes, we also generalize the target map.
 #ifdef DEBUG
@@ -3564,8 +3587,8 @@ RUNTIME_FUNCTION(Runtime_CloneObjectIC_Miss) {
               : GetCloneModeForMap(source_map, null_proto_literal, isolate);
       auto UpdateState = [&](Handle<Map> target_map) {
         UpdateNexus(target_map);
-        if (!null_proto_literal && !InReadOnlySpace(*source_map) &&
-            !InReadOnlySpace(*target_map)) {
+        if (CanCacheCloneTargetMapTransition(source_map, target_map,
+                                             null_proto_literal)) {
           SetCloneTargetMap(isolate, source_map, target_map);
         }
       };
@@ -3595,8 +3618,8 @@ RUNTIME_FUNCTION(Runtime_CloneObjectIC_Miss) {
                       result_map->GetInObjectProperties());
             UpdateState(result_map);
           } else {
-            if (!null_proto_literal && !InReadOnlySpace(*source_map) &&
-                !source_map->is_prototype_map()) {
+            if (CanCacheCloneTargetMapTransition(source_map, {},
+                                                 null_proto_literal)) {
               SetCloneTargetMapUnsupported(isolate, source_map);
             }
             if (nexus) {
diff --git a/src/objects/map-updater.cc b/src/objects/map-updater.cc
index 09fb7b6e14b..cae60e166e4 100644
--- a/src/objects/map-updater.cc
+++ b/src/objects/map-updater.cc
@@ -1210,8 +1210,7 @@ void MapUpdater::UpdateFieldType(Isolate* isolate, Handle<Map> map,
                                  InternalIndex descriptor, Handle<Name> name,
                                  PropertyConstness new_constness,
                                  Representation new_representation,
-                                 const MaybeObjectHandle& new_wrapped_type) {
-  DCHECK(IsSmi(*new_wrapped_type) || IsWeak(*new_wrapped_type));
+                                 Handle<FieldType> new_type) {
   // We store raw pointers in the queue, so no allocations are allowed.
   DisallowGarbageCollection no_gc;
   PropertyDetails details =
@@ -1250,24 +1249,48 @@ void MapUpdater::UpdateFieldType(Isolate* isolate, Handle<Map> map,
 #endif
         },
         TransitionsAccessor::IterationMode::kIncludeSideStepTransitions);
+
     Tagged<DescriptorArray> descriptors =
         current->instance_descriptors(isolate);
     details = descriptors->GetDetails(descriptor);
 
+    PropertyConstness cur_new_constness = new_constness;
+    Representation cur_new_representation = new_representation;
+    Handle<FieldType> cur_new_type = new_type;
+    // Through side-steps we can reach transition trees which are already more
+    // generalized. Ensure we don't re-concretize them.
+    if (!sidestep_transition.empty()) {
+      cur_new_constness =
+          GeneralizeConstness(new_constness, details.constness());
+      cur_new_representation =
+          new_representation.generalize(details.representation());
+      cur_new_type = GeneralizeFieldType(
+          details.representation(),
+          handle(descriptors->GetFieldType(descriptor), isolate),
+          cur_new_representation, new_type, isolate);
+      DCHECK(new_representation.fits_into(cur_new_representation));
+    }
+
     // It is allowed to change representation here only from None
     // to something or from Smi or HeapObject to Tagged.
-    DCHECK(details.representation().Equals(new_representation) ||
-           details.representation().CanBeInPlaceChangedTo(new_representation));
-
-    // Skip if already updated the shared descriptor.
-    if (new_constness != details.constness() ||
-        !new_representation.Equals(details.representation()) ||
-        descriptors->GetFieldType(descriptor) != *new_wrapped_type.object()) {
-      Descriptor d = Descriptor::DataField(
-          name, descriptors->GetFieldIndex(descriptor), details.attributes(),
-          new_constness, new_representation, new_wrapped_type);
-      descriptors->Replace(descriptor, &d);
+    CHECK(
+        details.representation().Equals(cur_new_representation) ||
+        details.representation().CanBeInPlaceChangedTo(cur_new_representation));
+
+    // Skip if we already updated the shared descriptor or the target was more
+    // general in the first place.
+    if (cur_new_constness == details.constness() &&
+        cur_new_representation.Equals(details.representation()) &&
+        FieldType::Equals(descriptors->GetFieldType(descriptor),
+                          *cur_new_type)) {
+      continue;
     }
+
+    MaybeObjectHandle wrapped_type(Map::WrapFieldType(new_type));
+    Descriptor d = Descriptor::DataField(
+        name, descriptors->GetFieldIndex(descriptor), details.attributes(),
+        cur_new_constness, cur_new_representation, wrapped_type);
+    descriptors->Replace(descriptor, &d);
   }
 }
 
@@ -1320,9 +1343,8 @@ void MapUpdater::GeneralizeField(Isolate* isolate, Handle<Map> map,
   PropertyDetails details = descriptors->GetDetails(modify_index);
   Handle<Name> name(descriptors->GetKey(modify_index), isolate);
 
-  MaybeObjectHandle wrapped_type(Map::WrapFieldType(new_field_type));
   UpdateFieldType(isolate, field_owner, modify_index, name, new_constness,
-                  new_representation, wrapped_type);
+                  new_representation, new_field_type);
 
   DCHECK_IMPLIES(IsClass(*new_field_type), new_representation.IsHeapObject());
 
diff --git a/src/objects/map-updater.h b/src/objects/map-updater.h
index 03ff806ac42..3aa948e766c 100644
--- a/src/objects/map-updater.h
+++ b/src/objects/map-updater.h
@@ -210,7 +210,7 @@ class V8_EXPORT_PRIVATE MapUpdater {
                               Handle<Name> name,
                               PropertyConstness new_constness,
                               Representation new_representation,
-                              const MaybeObjectHandle& new_wrapped_type);
+                              Handle<FieldType> new_type);
 
   void GeneralizeField(Handle<Map> map, InternalIndex modify_index,
                        PropertyConstness new_constness,
diff --git a/test/mjsunit/regress-343338105.js b/test/mjsunit/regress-343338105.js
new file mode 100644
index 00000000000..8ca0541de1e
--- /dev/null
+++ b/test/mjsunit/regress-343338105.js
@@ -0,0 +1,17 @@
+// Copyright 2024 the V8 project authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+function f2(a4) {
+  return {
+      ...a4,
+      "x": a4,
+      __proto__: a4,
+  };
+}
+const v14 = f2(f2);
+const o26 = {
+    ...v14,
+    "x": 2626,
+};
+f2(f2);
