diff --git a/src/compiler/backend/code-generator.cc b/src/compiler/backend/code-generator.cc
index 9e06d9e1614..117e9cba2c2 100644
--- a/src/compiler/backend/code-generator.cc
+++ b/src/compiler/backend/code-generator.cc
@@ -1290,7 +1290,9 @@ void CodeGenerator::BuildTranslationForFrameStateDescriptor(
       break;
     }
     case FrameStateType::kInlinedExtraArguments:
-      translations_.BeginInlinedExtraArguments(shared_info_id, height);
+      translations_.BeginInlinedExtraArguments(
+          shared_info_id, height,
+          descriptor->bytecode_array().ToHandleChecked()->parameter_count());
       break;
     case FrameStateType::kConstructCreateStub:
       translations_.BeginConstructCreateStubFrame(shared_info_id, height);
diff --git a/src/compiler/js-inlining.cc b/src/compiler/js-inlining.cc
index a222d3a9886..cec7300b5eb 100644
--- a/src/compiler/js-inlining.cc
+++ b/src/compiler/js-inlining.cc
@@ -249,15 +249,20 @@ Reduction JSInliner::InlineCall(Node* call, Node* new_target, Node* context,
 }
 
 FrameState JSInliner::CreateArtificialFrameState(
-    Node* node, FrameState outer_frame_state, int parameter_count,
+    Node* node, FrameState outer_frame_state, int argument_count,
     FrameStateType frame_state_type, SharedFunctionInfoRef shared,
-    Node* context, Node* callee) {
-  const int parameter_count_with_receiver =
-      parameter_count + JSCallOrConstructNode::kReceiverOrNewTargetInputCount;
+    OptionalBytecodeArrayRef maybe_bytecode_array, Node* context,
+    Node* callee) {
+  const int argument_count_with_receiver =
+      argument_count + JSCallOrConstructNode::kReceiverOrNewTargetInputCount;
+  IndirectHandle<BytecodeArray> bytecode_array_handle = {};
+  if (maybe_bytecode_array.has_value()) {
+    bytecode_array_handle = maybe_bytecode_array->object();
+  }
   const FrameStateFunctionInfo* state_info =
-      common()->CreateFrameStateFunctionInfo(frame_state_type,
-                                             parameter_count_with_receiver, 0,
-                                             0, shared.object(), {});
+      common()->CreateFrameStateFunctionInfo(
+          frame_state_type, argument_count_with_receiver, 0, 0, shared.object(),
+          bytecode_array_handle);
 
   const Operator* op = common()->FrameState(
       BytecodeOffset::None(), OutputFrameStateCombine::Ignore(), state_info);
@@ -286,7 +291,7 @@ FrameState JSInliner::CreateArtificialFrameState(
     NodeVector params(local_zone_);
     params.push_back(
         node->InputAt(JSCallOrConstructNode::ReceiverOrNewTargetIndex()));
-    for (int i = 0; i < parameter_count; i++) {
+    for (int i = 0; i < argument_count; i++) {
       params.push_back(node->InputAt(JSCallOrConstructNode::ArgumentIndex(i)));
     }
     const Operator* op_param = common()->StateValues(
@@ -604,7 +609,7 @@ void JSInliner::InlineWasmFunction(Node* call, Node* inlinee_start,
   Node* callee = jsgraph()->UndefinedConstant();
   Node* frame_state_inside = CreateArtificialFrameState(
       call, FrameState{frame_state}, argument_count,
-      FrameStateType::kWasmInlinedIntoJS, shared_fct_info, context, callee);
+      FrameStateType::kWasmInlinedIntoJS, shared_fct_info, {}, context, callee);
   Node* check_point = graph()->NewNode(common()->Checkpoint(),
                                        frame_state_inside, effect, control);
   effect = check_point;
@@ -863,7 +868,8 @@ Reduction JSInliner::ReduceJSCall(Node* node) {
       } else {
         frame_state_inside = CreateArtificialFrameState(
             node, frame_state, n.ArgumentCount(),
-            FrameStateType::kConstructCreateStub, *shared_info, caller_context);
+            FrameStateType::kConstructCreateStub, *shared_info, bytecode_array,
+            caller_context);
       }
       Node* create =
           graph()->NewNode(javascript()->Create(), call.target(), new_target,
@@ -917,7 +923,7 @@ Reduction JSInliner::ReduceJSCall(Node* node) {
     // reconstruct the proper frame when deoptimizing within the constructor.
     frame_state = CreateArtificialFrameState(
         node, frame_state, 0, FrameStateType::kConstructInvokeStub,
-        *shared_info, caller_context);
+        *shared_info, bytecode_array, caller_context);
   }
 
   // Insert a JSConvertReceiver node for sloppy callees. Note that the context
@@ -950,7 +956,7 @@ Reduction JSInliner::ReduceJSCall(Node* node) {
   if (call.argument_count() != parameter_count) {
     frame_state = CreateArtificialFrameState(
         node, frame_state, call.argument_count(),
-        FrameStateType::kInlinedExtraArguments, *shared_info);
+        FrameStateType::kInlinedExtraArguments, *shared_info, bytecode_array);
   }
 
   return InlineCall(node, new_target, context, frame_state, start, end,
diff --git a/src/compiler/js-inlining.h b/src/compiler/js-inlining.h
index 4356fe9be31..547b11bbd3b 100644
--- a/src/compiler/js-inlining.h
+++ b/src/compiler/js-inlining.h
@@ -111,7 +111,8 @@ class JSInliner final : public AdvancedReducer {
   FrameState CreateArtificialFrameState(
       Node* node, FrameState outer_frame_state, int parameter_count,
       FrameStateType frame_state_type, SharedFunctionInfoRef shared,
-      Node* context = nullptr, Node* callee = nullptr);
+      OptionalBytecodeArrayRef maybe_bytecode_array, Node* context = nullptr,
+      Node* callee = nullptr);
 
   Reduction InlineCall(Node* call, Node* new_target, Node* context,
                        Node* frame_state, StartNode start, Node* end,
diff --git a/src/compiler/turboshaft/maglev-graph-building-phase.cc b/src/compiler/turboshaft/maglev-graph-building-phase.cc
index 8a50252241b..7700b4dceec 100644
--- a/src/compiler/turboshaft/maglev-graph-building-phase.cc
+++ b/src/compiler/turboshaft/maglev-graph-building-phase.cc
@@ -4993,9 +4993,11 @@ class GraphBuildingNodeProcessor {
     int local_count = 0;
     Handle<SharedFunctionInfo> shared_info =
         maglev_frame.unit().shared_function_info().object();
+    Handle<BytecodeArray> bytecode_array =
+        maglev_frame.unit().bytecode().object();
     FrameStateFunctionInfo* info = graph_zone()->New<FrameStateFunctionInfo>(
         type, parameter_count, max_arguments, local_count, shared_info,
-        kNullMaybeHandle);
+        bytecode_array);
 
     return graph_zone()->New<FrameStateInfo>(maglev_frame.bytecode_position(),
                                              OutputFrameStateCombine::Ignore(),
@@ -5007,12 +5009,14 @@ class GraphBuildingNodeProcessor {
     FrameStateType type = FrameStateType::kConstructInvokeStub;
     Handle<SharedFunctionInfo> shared_info =
         maglev_frame.unit().shared_function_info().object();
+    Handle<BytecodeArray> bytecode_array =
+        maglev_frame.unit().bytecode().object();
     constexpr uint16_t kParameterCount = 1;  // Only 1 parameter: the receiver.
     constexpr uint16_t kMaxArguments = 0;
     constexpr int kLocalCount = 0;
     FrameStateFunctionInfo* info = graph_zone()->New<FrameStateFunctionInfo>(
         type, kParameterCount, kMaxArguments, kLocalCount, shared_info,
-        kNullMaybeHandle);
+        bytecode_array);
 
     return graph_zone()->New<FrameStateInfo>(
         BytecodeOffset::None(), OutputFrameStateCombine::Ignore(), info);
diff --git a/src/deoptimizer/deoptimizer.cc b/src/deoptimizer/deoptimizer.cc
index 91c5e42587b..df67a74a694 100644
--- a/src/deoptimizer/deoptimizer.cc
+++ b/src/deoptimizer/deoptimizer.cc
@@ -2101,15 +2101,17 @@ void Deoptimizer::DoComputeInlinedExtraArguments(
 
   TranslatedFrame::iterator value_iterator = translated_frame->begin();
   const int argument_count_without_receiver = translated_frame->height() - 1;
-  const int formal_parameter_count =
-      translated_frame->raw_shared_info()
-          ->internal_formal_parameter_count_without_receiver();
+  const int formal_parameter_count_without_receiver =
+      translated_frame->formal_parameter_count() - 1;
+  SBXCHECK_GE(formal_parameter_count_without_receiver, 0);
   const int extra_argument_count =
-      argument_count_without_receiver - formal_parameter_count;
+      argument_count_without_receiver - formal_parameter_count_without_receiver;
   // The number of pushed arguments is the maximum of the actual argument count
   // and the formal parameter count + the receiver.
-  const int padding = ArgumentPaddingSlots(
-      std::max(argument_count_without_receiver, formal_parameter_count) + 1);
+  const int padding =
+      ArgumentPaddingSlots(std::max(argument_count_without_receiver,
+                                    formal_parameter_count_without_receiver) +
+                           1);
   const int output_frame_size =
       (std::max(0, extra_argument_count) + padding) * kSystemPointerSize;
   if (verbose_tracing_enabled()) {
@@ -2146,7 +2148,8 @@ void Deoptimizer::DoComputeInlinedExtraArguments(
     // frame will do that.
     value_iterator++;  // Skip function.
     value_iterator++;  // Skip receiver.
-    for (int i = 0; i < formal_parameter_count; i++) value_iterator++;
+    for (int i = 0; i < formal_parameter_count_without_receiver; i++)
+      value_iterator++;
     frame_writer.PushStackJSArguments(value_iterator, extra_argument_count);
   }
 }
diff --git a/src/deoptimizer/frame-translation-builder.cc b/src/deoptimizer/frame-translation-builder.cc
index 73b054956fc..252bcf6a597 100644
--- a/src/deoptimizer/frame-translation-builder.cc
+++ b/src/deoptimizer/frame-translation-builder.cc
@@ -341,10 +341,11 @@ void FrameTranslationBuilder::BeginConstructInvokeStubFrame(int literal_id) {
   Add(opcode, SignedOperand(literal_id));
 }
 
-void FrameTranslationBuilder::BeginInlinedExtraArguments(int literal_id,
-                                                         unsigned height) {
+void FrameTranslationBuilder::BeginInlinedExtraArguments(
+    int literal_id, unsigned height, uint32_t parameter_count) {
   auto opcode = TranslationOpcode::INLINED_EXTRA_ARGUMENTS;
-  Add(opcode, SignedOperand(literal_id), UnsignedOperand(height));
+  Add(opcode, SignedOperand(literal_id), UnsignedOperand(height),
+      UnsignedOperand(parameter_count));
 }
 
 void FrameTranslationBuilder::BeginInterpretedFrame(
diff --git a/src/deoptimizer/frame-translation-builder.h b/src/deoptimizer/frame-translation-builder.h
index 15bedf4c583..eac969bba85 100644
--- a/src/deoptimizer/frame-translation-builder.h
+++ b/src/deoptimizer/frame-translation-builder.h
@@ -39,7 +39,8 @@ class FrameTranslationBuilder {
   void BeginInterpretedFrame(BytecodeOffset bytecode_offset, int literal_id,
                              int bytecode_array_id, unsigned height,
                              int return_value_offset, int return_value_count);
-  void BeginInlinedExtraArguments(int literal_id, unsigned height);
+  void BeginInlinedExtraArguments(int literal_id, unsigned height,
+                                  uint32_t parameter_count);
   void BeginConstructCreateStubFrame(int literal_id, unsigned height);
   void BeginConstructInvokeStubFrame(int literal_id);
   void BeginBuiltinContinuationFrame(BytecodeOffset bailout_id, int literal_id,
diff --git a/src/deoptimizer/translated-state.cc b/src/deoptimizer/translated-state.cc
index e721913e6b8..662f018168f 100644
--- a/src/deoptimizer/translated-state.cc
+++ b/src/deoptimizer/translated-state.cc
@@ -861,8 +861,11 @@ TranslatedFrame TranslatedFrame::UnoptimizedJSFrame(
 }
 
 TranslatedFrame TranslatedFrame::InlinedExtraArguments(
-    Tagged<SharedFunctionInfo> shared_info, uint32_t height) {
-  return TranslatedFrame(kInlinedExtraArguments, shared_info, {}, height);
+    Tagged<SharedFunctionInfo> shared_info, uint32_t height,
+    uint32_t formal_parameter_count) {
+  TranslatedFrame frame(kInlinedExtraArguments, shared_info, {}, height);
+  frame.formal_parameter_count_ = formal_parameter_count;
+  return frame;
 }
 
 TranslatedFrame TranslatedFrame::ConstructCreateStubFrame(
@@ -1069,12 +1072,15 @@ TranslatedFrame TranslatedState::CreateNextTranslatedFrame(
       Tagged<SharedFunctionInfo> shared_info = Cast<SharedFunctionInfo>(
           literal_array.get_on_heap_literals()->get(iterator->NextOperand()));
       uint32_t height = iterator->NextOperandUnsigned();
+      uint32_t parameter_count = iterator->NextOperandUnsigned();
       if (trace_file != nullptr) {
         std::unique_ptr<char[]> name = shared_info->DebugNameCStr();
         PrintF(trace_file, "  reading inlined arguments frame %s", name.get());
-        PrintF(trace_file, " => height=%u; inputs:\n", height);
+        PrintF(trace_file, " => height=%u, parameter_count=%u; inputs:\n",
+               height, parameter_count);
       }
-      return TranslatedFrame::InlinedExtraArguments(shared_info, height);
+      return TranslatedFrame::InlinedExtraArguments(shared_info, height,
+                                                    parameter_count);
     }
 
     case TranslationOpcode::CONSTRUCT_CREATE_STUB_FRAME: {
diff --git a/src/deoptimizer/translated-state.h b/src/deoptimizer/translated-state.h
index 9afadd3479c..e9c20c97f74 100644
--- a/src/deoptimizer/translated-state.h
+++ b/src/deoptimizer/translated-state.h
@@ -240,6 +240,10 @@ class TranslatedFrame {
 
   int return_value_offset() const { return return_value_offset_; }
   int return_value_count() const { return return_value_count_; }
+  int formal_parameter_count() const {
+    DCHECK_EQ(kind(), kInlinedExtraArguments);
+    return formal_parameter_count_;
+  }
 
   Tagged<SharedFunctionInfo> raw_shared_info() const {
     CHECK_EQ(handle_state_, kRawPointers);
@@ -326,7 +330,8 @@ class TranslatedFrame {
   static TranslatedFrame AccessorFrame(Kind kind,
                                        Tagged<SharedFunctionInfo> shared_info);
   static TranslatedFrame InlinedExtraArguments(
-      Tagged<SharedFunctionInfo> shared_info, uint32_t height);
+      Tagged<SharedFunctionInfo> shared_info, uint32_t height,
+      uint32_t formal_parameter_count);
   static TranslatedFrame ConstructCreateStubFrame(
       Tagged<SharedFunctionInfo> shared_info, uint32_t height);
   static TranslatedFrame ConstructInvokeStubFrame(
@@ -362,7 +367,6 @@ class TranslatedFrame {
                            uint32_t height, int return_value_offset = 0,
                            int return_value_count = 0)
       : kind_(kind),
-        bytecode_offset_(BytecodeOffset::None()),
         raw_shared_info_(raw_shared_info),
         raw_bytecode_array_(raw_bytecode_array),
         height_(height),
@@ -375,7 +379,7 @@ class TranslatedFrame {
   void Handlify(Isolate* isolate);
 
   Kind kind_;
-  BytecodeOffset bytecode_offset_;
+  BytecodeOffset bytecode_offset_ = BytecodeOffset::None();
 
   // Object references are stored as either raw pointers (before Handlify is
   // called) or handles (afterward).
@@ -391,6 +395,7 @@ class TranslatedFrame {
   uint32_t height_;
   int return_value_offset_;
   int return_value_count_;
+  int formal_parameter_count_ = -1;
 
   enum HandleState { kRawPointers, kHandles } handle_state_;
 
diff --git a/src/deoptimizer/translation-opcode.h b/src/deoptimizer/translation-opcode.h
index c36c287cf60..0bc7a1c202b 100644
--- a/src/deoptimizer/translation-opcode.h
+++ b/src/deoptimizer/translation-opcode.h
@@ -24,7 +24,7 @@ namespace internal {
   IF_WASM(V, JS_TO_WASM_BUILTIN_CONTINUATION_FRAME, 4) \
   IF_WASM(V, WASM_INLINED_INTO_JS_FRAME, 3)            \
   IF_WASM(V, LIFTOFF_FRAME, 3)                         \
-  V(INLINED_EXTRA_ARGUMENTS, 2)
+  V(INLINED_EXTRA_ARGUMENTS, 3)
 
 #define TRANSLATION_OPCODE_LIST(V)    \
   TRANSLATION_JS_FRAME_OPCODE_LIST(V) \
diff --git a/src/maglev/maglev-code-generator.cc b/src/maglev/maglev-code-generator.cc
index 8f93b7df429..333f705162c 100644
--- a/src/maglev/maglev-code-generator.cc
+++ b/src/maglev/maglev-code-generator.cc
@@ -1337,7 +1337,8 @@ class MaglevFrameTranslationBuilder {
                              const VirtualObject::List& virtual_objects) {
     translation_array_builder_->BeginInlinedExtraArguments(
         GetDeoptLiteral(GetSharedFunctionInfo(frame)),
-        static_cast<uint32_t>(frame.arguments().size()));
+        static_cast<uint32_t>(frame.arguments().size()),
+        GetBytecodeArray(frame).parameter_count());
 
     // Closure
     BuildDeoptFrameSingleValue(frame.closure(), current_input_location,
diff --git a/test/mjsunit/sandbox/regress/regress-379418918.js b/test/mjsunit/sandbox/regress/regress-379418918.js
new file mode 100644
index 00000000000..5140a67e688
--- /dev/null
+++ b/test/mjsunit/sandbox/regress/regress-379418918.js
@@ -0,0 +1,36 @@
+// Copyright 2024 the V8 project authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+//
+// Flags: --sandbox-fuzzing --allow-natives-syntax
+
+function corrupt(func) {
+  const kHeapObjectTagMask = 0x3;
+  const kJSFunctionType = Sandbox.getInstanceTypeIdFor('JS_FUNCTION_TYPE');
+  const kSharedFunctionInfoOffset =
+      Sandbox.getFieldOffset(kJSFunctionType, 'shared_function_info');
+  const kSharedFunctionInfoType =
+      Sandbox.getInstanceTypeIdFor('SHARED_FUNCTION_INFO_TYPE');
+  const kFormalParameterCountOffset =
+      Sandbox.getFieldOffset(kSharedFunctionInfoType, 'formal_parameter_count');
+
+  const memview = new DataView(new Sandbox.MemoryView(0, 0x100000000));
+  let func_addr = Sandbox.getAddressOf(func);
+  let sfi_addr = memview.getUint32(func_addr + kSharedFunctionInfoOffset, true) & ~kHeapObjectTagMask;
+  memview.setUint16(sfi_addr + kFormalParameterCountOffset, 32, true);
+}
+
+function f0(acc, value) {
+  value.b = value;
+}
+corrupt(f0);
+
+function f3() {
+  const v5 = Array(Array);
+  v5.reduce(f0, 0);
+}
+%PrepareFunctionForOptimization(f3);
+%PrepareFunctionForOptimization(f0);
+f3();
+%OptimizeFunctionOnNextCall(f3);
+f3();
